
\begin{DoxyItemize}
\item \hyperlink{newAPI}{R\+N\+Alib A\+PI v3.\+0}
\item \hyperlink{callbacks}{Callback Functions}
\item \hyperlink{wrappers}{Scripting Language interface(s)} 
\end{DoxyItemize}\hypertarget{newAPI}{}\section{R\+N\+Alib A\+PI v3.0}\label{newAPI}
\hypertarget{newAPI_newAPI_intro}{}\subsection{Introduction}\label{newAPI_newAPI_intro}
With version 2.\+2 we introduce the new A\+PI that will take over the old one in the future version 3.\+0. By then, backwards compatibility will be broken, and third party applications using R\+N\+Alib need to be ported. This switch of A\+PI became necessary, since many new features found their way into the R\+N\+Alib where a balance between threadsafety and easy-\/to-\/use library functions is hard or even impossible to establish. Furthermore, many old functions of the library are present as slightly modified copies of themself to provide a crude way to overload functions.

Therefore, we introduce the new v3.\+0 A\+PI very early in our development stage such that developers have enough time to migrate to the new functions and interfaces. We also started to provide encapsulation of the R\+N\+Alib functions, data structures, typedefs, and macros by prefixing them with {\itshape vrna\+\_\+} and {\itshape V\+R\+N\+A\+\_\+} , respectively. Header files should also be included using the {\itshape Vienna\+R\+N\+A/} namespace, e.\+g. 
\begin{DoxyCode}
\textcolor{preprocessor}{#include <\hyperlink{fold_8h}{ViennaRNA/fold.h}>}
\end{DoxyCode}
 instead of just using 
\begin{DoxyCode}
\textcolor{preprocessor}{#include <\hyperlink{fold_8h}{fold.h}>}
\end{DoxyCode}
 as required for R\+N\+Alib 1.\+x and 2.\+x.

This eases the work for programmers of third party applications that would otherwise need to put much effort into renaming functions and data types in their own implementations if their names appear in our library. Since we still provide backward compatibility up to the last version of R\+N\+Alib 2.\+x, this advantage may be fully exploited only starting from v3.\+0 which will be released in the future. However, our plan is to provide the possibility for an early switch-\/off mechanism of the backward compatibility in one of our next releases of Vienna\+R\+NA Package 2.\+x.\hypertarget{newAPI_newAPI_changes}{}\subsection{What are the major changes?}\label{newAPI_newAPI_changes}
...\hypertarget{newAPI_newAPI_porting}{}\subsection{How to port your program to the new A\+PI}\label{newAPI_newAPI_porting}
...\hypertarget{newAPI_newAPI_examples}{}\subsection{Some Examples using R\+N\+Alib A\+P\+I v3.\+0}\label{newAPI_newAPI_examples}
Examples on how to use the new v3.\+0 A\+PI can be found in the examples\+\_\+c\+\_\+new\+\_\+\+A\+PI section. \hypertarget{callbacks}{}\section{Callback Functions}\label{callbacks}
With the new \hyperlink{newAPI}{R\+N\+Alib A\+PI v3.\+0} we introduce so-\/called callback mechanisms for several functions.\hypertarget{callbacks_callbacks_intro}{}\subsection{The purpose of Callback mechanisms}\label{callbacks_callbacks_intro}
Using callback mechanisms, our library enables users not only to retrieve computed data without the need for parsing complicated data structures, but also allows one to tweak our implementation to do additional tasks without the requirement of a re-\/implementation of basic algorithms.

Our implementation of the callback mechanisms always follows the same scheme\+: The user\+:
\begin{DoxyItemize}
\item defines a function that complies with the interface we\textquotesingle{}ve defined, and
\item passes a pointer to said function to our implementations
\end{DoxyItemize}

In addition to the specific arguments of our callback interfaces, virtually all callbacks receive an additional {\itshape pass-\/through-\/pointer} as their last argument. This enables one to\+:
\begin{DoxyItemize}
\item encapsulate data, and
\item provide thread-\/safe operations,
\end{DoxyItemize}

since this pointer is simply passed through by our library functions. It may therefore hold the address of an arbitrary, user-\/defined data structure.\hypertarget{callbacks_callback_descriptions}{}\subsection{List of available Callbacks}\label{callbacks_callback_descriptions}
Below, you find an enumeration of the individual callback functions that are available in {\itshape R\+N\+Alib}.


\begin{DoxyRefList}
\item[\label{callbacks__callbacks000005}%
\Hypertarget{callbacks__callbacks000005}%
Global \hyperlink{group__fold__compound_ga7806651f51b195013839a218b3bbd5a3}{vrna\+\_\+callback\+\_\+free\+\_\+auxdata} (void $\ast$data)]This callback is supposed to free memory occupied by an auxiliary data structure. It will be called when the \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} is erased from memory through a call to \hyperlink{group__fold__compound_gadded6039d63f5d6509836e20321534ad}{vrna\+\_\+fold\+\_\+compound\+\_\+free()} and will be passed the address of memory previously bound to the \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} via \hyperlink{group__fold__compound_ga6316a9426bea2f742375e8df6febd3f6}{vrna\+\_\+fold\+\_\+compound\+\_\+add\+\_\+auxdata()}.  
\item[\label{callbacks__callbacks000001}%
\Hypertarget{callbacks__callbacks000001}%
Global \hyperlink{group__hard__constraints_gae465f1d4a3d8b6592b38ecbb0d9f613d}{vrna\+\_\+callback\+\_\+hc\+\_\+evaluate} (int i, int j, int k, int l, unsigned char d, void $\ast$data)]This callback enables one to over-\/rule default hard constraints in secondary structure decompositions.  
\item[\label{callbacks__callbacks000006}%
\Hypertarget{callbacks__callbacks000006}%
Global \hyperlink{group__fold__compound_gac86036fa8cad1108832335063243cdc8}{vrna\+\_\+callback\+\_\+recursion\+\_\+status} (unsigned char status, void $\ast$data)]This function will be called to notify a third-\/party implementation about the status of a currently ongoing recursion. The purpose of this callback mechanism is to provide users with a simple way to ensure pre-\/ and post conditions for auxiliary mechanisms attached to our implementations.  
\item[\label{callbacks__callbacks000004}%
\Hypertarget{callbacks__callbacks000004}%
Global \hyperlink{group__soft__constraints_gaeb6448da6c593d4c489c7fbadcb99499}{vrna\+\_\+callback\+\_\+sc\+\_\+backtrack} (int i, int j, int k, int l, unsigned char d, void $\ast$data)]This callback enables one to add auxiliary base pairs in the backtracking steps of hairpin-\/ and interior loops.  
\item[\label{callbacks__callbacks000002}%
\Hypertarget{callbacks__callbacks000002}%
Global \hyperlink{group__soft__constraints_ga88a266695d9e25cc12114dceb7b4565e}{vrna\+\_\+callback\+\_\+sc\+\_\+energy} (int i, int j, int k, int l, unsigned char d, void $\ast$data)]This callback enables one to add (pseudo-\/)energy contributions to individual decompositions of the secondary structure.  
\item[\label{callbacks__callbacks000003}%
\Hypertarget{callbacks__callbacks000003}%
Global \hyperlink{group__soft__constraints_ga4099978d410513edeeff8f3db13144c5}{vrna\+\_\+callback\+\_\+sc\+\_\+exp\+\_\+energy} (int i, int j, int k, int l, unsigned char d, void $\ast$data)]This callback enables one to add (pseudo-\/)energy contributions to individual decompositions of the secondary structure (Partition function variant, i.\+e. contributions must be returned as Boltzmann factors).  
\item[\label{callbacks__callbacks000010}%
\Hypertarget{callbacks__callbacks000010}%
Global \hyperlink{group__domains__up_ga75825c57d0bfde4ae4f95c044260c5c3}{vrna\+\_\+callback\+\_\+ud\+\_\+energy} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, int j, unsigned int loop\+\_\+type, void $\ast$data)]This function will be called to determine the additional energy contribution of a specific unstructured domain, e.\+g. the binding free energy of some ligand.  
\item[\label{callbacks__callbacks000011}%
\Hypertarget{callbacks__callbacks000011}%
Global \hyperlink{group__domains__up_ga861706f257ba993753464b823e65b86e}{vrna\+\_\+callback\+\_\+ud\+\_\+exp\+\_\+energy} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, int j, unsigned int loop\+\_\+type, void $\ast$data)]This function will be called to determine the additional energy contribution of a specific unstructured domain, e.\+g. the binding free energy of some ligand (Partition function variant, i.\+e. the Boltzmann factors instead of actual free energies).  
\item[\label{callbacks__callbacks000013}%
\Hypertarget{callbacks__callbacks000013}%
Global \hyperlink{group__domains__up_ga33d78327dcd04c1ca5ab2887edc18c7b}{vrna\+\_\+callback\+\_\+ud\+\_\+exp\+\_\+production} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, void $\ast$data)]The production rule for the unstructured domain grammar extension (Partition function variant)  
\item[\label{callbacks__callbacks000014}%
\Hypertarget{callbacks__callbacks000014}%
Global \hyperlink{group__domains__up_gab10498abc84fcaf336aca8f8d7d42eb2}{vrna\+\_\+callback\+\_\+ud\+\_\+probs\+\_\+add} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, int j, unsigned int loop\+\_\+type, F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+BL exp\+\_\+energy, void $\ast$data)]A callback function to store equilibrium probabilities for the unstructured domain feature  
\item[\label{callbacks__callbacks000015}%
\Hypertarget{callbacks__callbacks000015}%
Global \hyperlink{group__domains__up_gaa10ba1b6f1e179ea84c5caca9cdaae67}{vrna\+\_\+callback\+\_\+ud\+\_\+probs\+\_\+get} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, int j, unsigned int loop\+\_\+type, int motif, void $\ast$data)]A callback function to retrieve equilibrium probabilities for the unstructured domain feature  
\item[\label{callbacks__callbacks000012}%
\Hypertarget{callbacks__callbacks000012}%
Global \hyperlink{group__domains__up_ga4fdfc02c1b660c07f2d887772f02a0a1}{vrna\+\_\+callback\+\_\+ud\+\_\+production} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, void $\ast$data)]The production rule for the unstructured domain grammar extension  
\item[\label{callbacks__callbacks000008}%
\Hypertarget{callbacks__callbacks000008}%
Global \hyperlink{group__mfe__fold_ga4f3e5bc214ef803074ace313cb9571b4}{vrna\+\_\+mfe\+\_\+window\+\_\+callback} (int start, int end, const char $\ast$structure, float en, void $\ast$data)]This function will be called for each hit in a sliding window M\+FE prediction.  
\item[\label{callbacks__callbacks000007}%
\Hypertarget{callbacks__callbacks000007}%
Global \hyperlink{LPfold_8h_abe710a1182e6db69cc75329dfc9bed67}{vrna\+\_\+probs\+\_\+window\+\_\+callback} (F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+BL $\ast$pr, int pr\+\_\+size, int i, int max, unsigned int type, void $\ast$data)]This function will be called for each probability data set in the sliding window probability computation implementation of \hyperlink{group__local__pf__fold_ga70ec33810fa17a9f67a7c59f156d0137}{vrna\+\_\+probs\+\_\+window()}. The argument {\itshape type} specifies the type of probability that is passed to this function.  
\item[\label{callbacks__callbacks000009}%
\Hypertarget{callbacks__callbacks000009}%
Global \hyperlink{group__subopt__wuchty_gaa0270c66d04f59e750401695b8282e04}{vrna\+\_\+subopt\+\_\+callback} (const char $\ast$stucture, float energy, void $\ast$data)]This function will be called for each suboptimal secondary structure that is successfully backtraced. 
\end{DoxyRefList}\hypertarget{wrappers}{}\section{Scripting Language interface(s)}\label{wrappers}
\hypertarget{wrappers_scripting_intro}{}\subsection{Introduction}\label{wrappers_scripting_intro}
For an easy integration into scripting languages, we provide an automatically generated interface to the R\+N\+Alib C-\/library, generated with S\+W\+IG.\hypertarget{wrappers_scripting_renaming}{}\subsection{Function renaming scheme}\label{wrappers_scripting_renaming}
The main difference when using a scripting language interface compared to direct calls of R\+N\+Alib C functions is, that the prefix \textquotesingle{}vrna\+\_\+\textquotesingle{} is dropped. For instance, when calling the \hyperlink{group__mfe__fold__single_ga29a33b2895f4e67b0480271ff289afdc}{vrna\+\_\+fold()} function, corresponding calls in perl or python are R\+N\+A\+::fold(), and R\+N\+A.\+fold(), respectively.

Functions that are dedicated to work on specific data structures only, e.\+g. the \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t}, are usually not exported at all. Instead, they are attached as object methods of a corresponding class (see \hyperlink{wrappers_scripting_oo_interface}{Object oriented Interface for data structures} for detailed information).\hypertarget{wrappers_scripting_oo_interface}{}\subsection{Object oriented Interface for data structures}\label{wrappers_scripting_oo_interface}
For data structures, typedefs, and enumerations the \textquotesingle{}vrna\+\_\+\textquotesingle{} prefixes are dropped as well, together with their suffixes \textquotesingle{}\+\_\+s\textquotesingle{}, \textquotesingle{}\+\_\+t\textquotesingle{}, and \textquotesingle{}\+\_\+e\textquotesingle{}, respectively. Furthermore, data structures are usually transformed into classes and relevant functions of the C-\/library are attached as methods.\hypertarget{wrappers_scripting_examples}{}\subsection{Examples}\label{wrappers_scripting_examples}
Examples on the basic usage of the scripting language interfaces can be found in the scripting\+\_\+perl\+\_\+examples and scripting\+\_\+python\+\_\+examples section.\hypertarget{wrappers_scripting_wrapper_notes}{}\subsection{S\+W\+I\+G generated Wrapper notes}\label{wrappers_scripting_wrapper_notes}
Special notes on how functions, structures, enums, and macro definitions are actually wrapped, can be found below


\begin{DoxyRefList}
\item[\label{wrappers__wrappers000002}%
\Hypertarget{wrappers__wrappers000002}%
Global \hyperlink{group__aln__utils_gaa12b481a7e7b965ef2eb1bcc4399e759}{vrna\+\_\+aln\+\_\+conservation\+\_\+col} (const char $\ast$$\ast$alignment, const vrna\+\_\+md\+\_\+t $\ast$md\+\_\+p, unsigned int options)]This function is available in an overloaded form where the last two parameters may be omitted, indicating {\ttfamily md} = {\itshape N\+U\+LL}, and {\ttfamily options} = \hyperlink{group__aln__utils_ga1e659227c9fc077d29989f576f129000}{V\+R\+N\+A\+\_\+\+M\+E\+A\+S\+U\+R\+E\+\_\+\+S\+H\+A\+N\+N\+O\+N\+\_\+\+E\+N\+T\+R\+O\+PY}, respectively.  
\item[\label{wrappers__wrappers000001}%
\Hypertarget{wrappers__wrappers000001}%
Global \hyperlink{group__aln__utils_gab6f16a2ea93f3bfd4d089cc8d448bb16}{vrna\+\_\+aln\+\_\+conservation\+\_\+struct} (const char $\ast$$\ast$alignment, const char $\ast$structure, const vrna\+\_\+md\+\_\+t $\ast$md)]This function is available in an overloaded form where the last parameter may be omitted, indicating {\ttfamily md} = {\itshape N\+U\+LL}  
\item[\label{wrappers__wrappers000005}%
\Hypertarget{wrappers__wrappers000005}%
Global \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+cmd\+\_\+t $\ast$commands, unsigned int options)]This function is attached as method {\bfseries commands\+\_\+apply()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000088}%
\Hypertarget{wrappers__wrappers000088}%
Global \hyperlink{group__struct__utils_gae966b9f44168a4f4b39ca42ffb5f37b7}{vrna\+\_\+db\+\_\+flatten} (char $\ast$structure, unsigned int options)]This function flattens an input structure string in-\/place! The second parameter is optional and defaults to \hyperlink{group__struct__utils_ga559ebf76b1b289f85309f4206e99aa1a}{V\+R\+N\+A\+\_\+\+B\+R\+A\+C\+K\+E\+T\+S\+\_\+\+D\+E\+F\+A\+U\+LT}.

An overloaded version of this function exists, where an additional second parameter can be passed to specify the target brackets, i.\+e. the type of matching pair characters all brackets will be flattened to. Therefore, in the scripting language interface this function is a replacement for \hyperlink{group__struct__utils_ga690425199c8b71545e7196e3af1436f8}{vrna\+\_\+db\+\_\+flatten\+\_\+to()}.  
\item[\label{wrappers__wrappers000089}%
\Hypertarget{wrappers__wrappers000089}%
Global \hyperlink{group__struct__utils_ga690425199c8b71545e7196e3af1436f8}{vrna\+\_\+db\+\_\+flatten\+\_\+to} (char $\ast$string, const char target\mbox{[}3\mbox{]}, unsigned int options)]This function is available as an overloaded version of \hyperlink{group__struct__utils_gae966b9f44168a4f4b39ca42ffb5f37b7}{vrna\+\_\+db\+\_\+flatten()}  
\item[\label{wrappers__wrappers000003}%
\Hypertarget{wrappers__wrappers000003}%
Global \hyperlink{group__utils_gae081ac655a76bd5c4b3d86c60b096b75}{vrna\+\_\+enumerate\+\_\+necklaces} (const unsigned int $\ast$type\+\_\+counts)]This function is available as global function {\bfseries enumerate\+\_\+necklaces()} which accepts lists input, an produces list of lists output.  
\item[\label{wrappers__wrappers000040}%
\Hypertarget{wrappers__wrappers000040}%
Global \hyperlink{group__eval_gac96577cf232c71160f762737a994b7c6}{vrna\+\_\+eval\+\_\+circ\+\_\+consensus\+\_\+structure} (const char $\ast$$\ast$alignment, const char $\ast$structure)]This function is available through an overloadeded version of \hyperlink{group__eval_ga3e05a23ddf9b083f4e69881e440d4866}{vrna\+\_\+eval\+\_\+circ\+\_\+structure()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument  
\item[\label{wrappers__wrappers000050}%
\Hypertarget{wrappers__wrappers000050}%
Global \hyperlink{group__eval_gae89240c230e4740b22a703ee953396b9}{vrna\+\_\+eval\+\_\+circ\+\_\+consensus\+\_\+structure\+\_\+v} (const char $\ast$$\ast$alignment, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga3e05a23ddf9b083f4e69881e440d4866}{vrna\+\_\+eval\+\_\+circ\+\_\+structure()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000042}%
\Hypertarget{wrappers__wrappers000042}%
Global \hyperlink{group__eval_gac673ebb9ae2a29f54d201e2ac5b85540}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+consensus\+\_\+structure} (const char $\ast$$\ast$alignment, const char $\ast$structure)]This function is available through an overloadeded version of \hyperlink{group__eval_ga9dba2fc5d7e6ad1359a7c2f350589c0e}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+structure()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument  
\item[\label{wrappers__wrappers000052}%
\Hypertarget{wrappers__wrappers000052}%
Global \hyperlink{group__eval_gaecd3e17292a0b3927277434019a5e187}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+consensus\+\_\+structure\+\_\+v} (const char $\ast$$\ast$alignment, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga9dba2fc5d7e6ad1359a7c2f350589c0e}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+structure()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000038}%
\Hypertarget{wrappers__wrappers000038}%
Global \hyperlink{group__eval_ga9dba2fc5d7e6ad1359a7c2f350589c0e}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+structure} (const char $\ast$string, const char $\ast$structure)]In the target scripting language, this function serves as a wrapper for \hyperlink{group__eval_gab96a6c59923ff06c35f8c2fd2c239727}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+structure\+\_\+v()} and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000048}%
\Hypertarget{wrappers__wrappers000048}%
Global \hyperlink{group__eval_gab96a6c59923ff06c35f8c2fd2c239727}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+structure\+\_\+v} (const char $\ast$string, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga9dba2fc5d7e6ad1359a7c2f350589c0e}{vrna\+\_\+eval\+\_\+circ\+\_\+gquad\+\_\+structure()}. The last two arguments for this function are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000036}%
\Hypertarget{wrappers__wrappers000036}%
Global \hyperlink{group__eval_ga3e05a23ddf9b083f4e69881e440d4866}{vrna\+\_\+eval\+\_\+circ\+\_\+structure} (const char $\ast$string, const char $\ast$structure)]In the target scripting language, this function serves as a wrapper for \hyperlink{group__eval_gac3fb44e0773a51be8efc5f4f595a94a7}{vrna\+\_\+eval\+\_\+circ\+\_\+structure\+\_\+v()} and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000046}%
\Hypertarget{wrappers__wrappers000046}%
Global \hyperlink{group__eval_gac3fb44e0773a51be8efc5f4f595a94a7}{vrna\+\_\+eval\+\_\+circ\+\_\+structure\+\_\+v} (const char $\ast$string, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga3e05a23ddf9b083f4e69881e440d4866}{vrna\+\_\+eval\+\_\+circ\+\_\+structure()}. The last two arguments for this function are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000054}%
\Hypertarget{wrappers__wrappers000054}%
Global \hyperlink{group__eval_gabbb4d2a7aa324ec9cce8f47ce61ab8af}{vrna\+\_\+eval\+\_\+consensus\+\_\+structure\+\_\+pt\+\_\+simple} (const char $\ast$$\ast$alignment, const short $\ast$pt)]This function is available through an overloadeded version of \hyperlink{group__eval_ga0bba59b4d6e53461088666ff4aece7b0}{vrna\+\_\+eval\+\_\+structure\+\_\+pt\+\_\+simple()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument  
\item[\label{wrappers__wrappers000039}%
\Hypertarget{wrappers__wrappers000039}%
Global \hyperlink{group__eval_ga7762c3a7bdcbc3a14ef93259d322c7d6}{vrna\+\_\+eval\+\_\+consensus\+\_\+structure\+\_\+simple} (const char $\ast$$\ast$alignment, const char $\ast$structure)]This function is available through an overloadeded version of \hyperlink{group__eval_ga7e5273464b775d4130245681312c1369}{vrna\+\_\+eval\+\_\+structure\+\_\+simple()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument  
\item[\label{wrappers__wrappers000049}%
\Hypertarget{wrappers__wrappers000049}%
Global \hyperlink{group__eval_gad88927c62ab0a8b534e078e44be1b36e}{vrna\+\_\+eval\+\_\+consensus\+\_\+structure\+\_\+simple\+\_\+v} (const char $\ast$$\ast$alignment, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga7e5273464b775d4130245681312c1369}{vrna\+\_\+eval\+\_\+structure\+\_\+simple()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000044}%
\Hypertarget{wrappers__wrappers000044}%
Global \hyperlink{group__eval_ga1c07851f6b665c3461a19e9e4eb33d26}{vrna\+\_\+eval\+\_\+consensus\+\_\+structure\+\_\+simple\+\_\+verbose} (const char $\ast$$\ast$alignment, const char $\ast$structure, F\+I\+LE $\ast$file)]This function is not available. Use \hyperlink{group__eval_gad88927c62ab0a8b534e078e44be1b36e}{vrna\+\_\+eval\+\_\+consensus\+\_\+structure\+\_\+simple\+\_\+v()} instead!  
\item[\label{wrappers__wrappers000029}%
\Hypertarget{wrappers__wrappers000029}%
Global \hyperlink{group__eval_ga6cea75c0eb9857fb59172be54cab09e0}{vrna\+\_\+eval\+\_\+covar\+\_\+structure} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$structure)]This function is attached as method {\bfseries eval\+\_\+covar\+\_\+structure()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000041}%
\Hypertarget{wrappers__wrappers000041}%
Global \hyperlink{group__eval_gaf09a326b3d57a4b30c27bd0e216198ac}{vrna\+\_\+eval\+\_\+gquad\+\_\+consensus\+\_\+structure} (const char $\ast$$\ast$alignment, const char $\ast$structure)]This function is available through an overloadeded version of \hyperlink{group__eval_ga3263504825ef4b523eba797c99921df4}{vrna\+\_\+eval\+\_\+gquad\+\_\+structure()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument  
\item[\label{wrappers__wrappers000051}%
\Hypertarget{wrappers__wrappers000051}%
Global \hyperlink{group__eval_ga8abc794fc48d43268ced5e8cde017baa}{vrna\+\_\+eval\+\_\+gquad\+\_\+consensus\+\_\+structure\+\_\+v} (const char $\ast$$\ast$alignment, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga3263504825ef4b523eba797c99921df4}{vrna\+\_\+eval\+\_\+gquad\+\_\+structure()}. Simply pass a sequence alignment as list of strings (including gaps) as first, and the consensus structure as second argument. The last two arguments are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000037}%
\Hypertarget{wrappers__wrappers000037}%
Global \hyperlink{group__eval_ga3263504825ef4b523eba797c99921df4}{vrna\+\_\+eval\+\_\+gquad\+\_\+structure} (const char $\ast$string, const char $\ast$structure)]In the target scripting language, this function serves as a wrapper for \hyperlink{group__eval_gaeaa2bdbc1b5d78c667e735fbdff87fff}{vrna\+\_\+eval\+\_\+gquad\+\_\+structure\+\_\+v()} and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000047}%
\Hypertarget{wrappers__wrappers000047}%
Global \hyperlink{group__eval_gaeaa2bdbc1b5d78c667e735fbdff87fff}{vrna\+\_\+eval\+\_\+gquad\+\_\+structure\+\_\+v} (const char $\ast$string, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga3263504825ef4b523eba797c99921df4}{vrna\+\_\+eval\+\_\+gquad\+\_\+structure()}. The last two arguments for this function are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000033}%
\Hypertarget{wrappers__wrappers000033}%
Global \hyperlink{group__loops_gab3eb4651dc26dc2b653a57dd340d7e68}{vrna\+\_\+eval\+\_\+hp\+\_\+loop} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, int j)]This function is attached as method {\bfseries eval\+\_\+hp\+\_\+loop()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000034}%
\Hypertarget{wrappers__wrappers000034}%
Global \hyperlink{group__loops_ga018f1314dbbae42fdd27c94670b61721}{vrna\+\_\+eval\+\_\+int\+\_\+loop} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, int j, int k, int l)]This function is attached as method {\bfseries eval\+\_\+int\+\_\+loop()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000030}%
\Hypertarget{wrappers__wrappers000030}%
Global \hyperlink{group__eval_ga730ba4df55c02fd530a0cddd49faf760}{vrna\+\_\+eval\+\_\+loop\+\_\+pt} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, const short $\ast$pt)]This function is attached as method {\bfseries eval\+\_\+loop\+\_\+pt()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000031}%
\Hypertarget{wrappers__wrappers000031}%
Global \hyperlink{group__eval_gaff1b9e4f4d17b434b0a822fe783672c1}{vrna\+\_\+eval\+\_\+move} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$structure, int m1, int m2)]This function is attached as method {\bfseries eval\+\_\+move()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000032}%
\Hypertarget{wrappers__wrappers000032}%
Global \hyperlink{group__eval_ga123dabc119ea98c968a5e903cc46f0fb}{vrna\+\_\+eval\+\_\+move\+\_\+pt} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, short $\ast$pt, int m1, int m2)]This function is attached as method {\bfseries eval\+\_\+move\+\_\+pt()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000025}%
\Hypertarget{wrappers__wrappers000025}%
Global \hyperlink{group__eval_ga58f199f1438d794a265f3b27fc8ea631}{vrna\+\_\+eval\+\_\+structure} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$structure)]This function is attached as method {\bfseries eval\+\_\+structure()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000026}%
\Hypertarget{wrappers__wrappers000026}%
Global \hyperlink{group__eval_gadbd09372ddfd7a450bbd590c96a6bfe4}{vrna\+\_\+eval\+\_\+structure\+\_\+pt} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const short $\ast$pt)]This function is attached as method {\bfseries eval\+\_\+structure\+\_\+pt()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000053}%
\Hypertarget{wrappers__wrappers000053}%
Global \hyperlink{group__eval_ga0bba59b4d6e53461088666ff4aece7b0}{vrna\+\_\+eval\+\_\+structure\+\_\+pt\+\_\+simple} (const char $\ast$string, const short $\ast$pt)]In the target scripting language, this function serves as a wrapper for \hyperlink{group__eval_ga2c6533ba0afe4c88d335d8f1e0e2a48e}{vrna\+\_\+eval\+\_\+structure\+\_\+pt\+\_\+v()} and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000028}%
\Hypertarget{wrappers__wrappers000028}%
Global \hyperlink{group__eval_ga8a517cfeeae8c376ae7b1e0c401d38b4}{vrna\+\_\+eval\+\_\+structure\+\_\+pt\+\_\+verbose} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const short $\ast$pt, F\+I\+LE $\ast$file)]This function is attached as method {\bfseries eval\+\_\+structure\+\_\+pt\+\_\+verbose()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000035}%
\Hypertarget{wrappers__wrappers000035}%
Global \hyperlink{group__eval_ga7e5273464b775d4130245681312c1369}{vrna\+\_\+eval\+\_\+structure\+\_\+simple} (const char $\ast$string, const char $\ast$structure)]In the target scripting language, this function serves as a wrapper for \hyperlink{group__eval_gacd6278343e77d13f1d53588e50d303bc}{vrna\+\_\+eval\+\_\+structure\+\_\+simple\+\_\+v()} and, thus, allows for two additional, optional arguments, the verbosity level and a file handle which default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000045}%
\Hypertarget{wrappers__wrappers000045}%
Global \hyperlink{group__eval_gacd6278343e77d13f1d53588e50d303bc}{vrna\+\_\+eval\+\_\+structure\+\_\+simple\+\_\+v} (const char $\ast$string, const char $\ast$structure, int verbosity\+\_\+level, F\+I\+LE $\ast$file)]This function is available through an overloaded version of \hyperlink{group__eval_ga7e5273464b775d4130245681312c1369}{vrna\+\_\+eval\+\_\+structure\+\_\+simple()}. The last two arguments for this function are optional and default to \#\+V\+R\+N\+A\+\_\+\+V\+E\+R\+B\+O\+S\+I\+T\+Y\+\_\+\+Q\+U\+I\+ET and N\+U\+LL, respectively.  
\item[\label{wrappers__wrappers000043}%
\Hypertarget{wrappers__wrappers000043}%
Global \hyperlink{group__eval_gaf928bfd96767e1b8033a95a4cc432e39}{vrna\+\_\+eval\+\_\+structure\+\_\+simple\+\_\+verbose} (const char $\ast$string, const char $\ast$structure, F\+I\+LE $\ast$file)]This function is not available. Use \hyperlink{group__eval_gacd6278343e77d13f1d53588e50d303bc}{vrna\+\_\+eval\+\_\+structure\+\_\+simple\+\_\+v()} instead!  
\item[\label{wrappers__wrappers000027}%
\Hypertarget{wrappers__wrappers000027}%
Global \hyperlink{group__eval_ga0928d699d310178f84ee2351034e5cb5}{vrna\+\_\+eval\+\_\+structure\+\_\+verbose} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$structure, F\+I\+LE $\ast$file)]This function is attached as method {\bfseries eval\+\_\+structure\+\_\+verbose()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000080}%
\Hypertarget{wrappers__wrappers000080}%
Global \hyperlink{group__energy__parameters_gad607bc3a5b5da16400e2ca4ea5560233}{vrna\+\_\+exp\+\_\+params\+\_\+rescale} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, double $\ast$mfe)]This function is attached to \hyperlink{group__fold__compound_structvrna__fc__s}{vrna\+\_\+fc\+\_\+s} objects as overloaded {\bfseries exp\+\_\+params\+\_\+rescale()} method.

When no parameter is passed to this method, the resulting action is the same as passing {\itshape N\+U\+LL} as second parameter to \hyperlink{group__energy__parameters_gad607bc3a5b5da16400e2ca4ea5560233}{vrna\+\_\+exp\+\_\+params\+\_\+rescale()}, i.\+e. default scaling of the partition function. Passing an energy in kcal/mol, e.\+g. as retrieved by a previous call to the {\itshape mfe()} method, instructs all subsequent calls to scale the partition function accordingly.  
\item[\label{wrappers__wrappers000082}%
\Hypertarget{wrappers__wrappers000082}%
Global \hyperlink{group__energy__parameters_gaa5409218068be84d7b50c78fbdaa85a9}{vrna\+\_\+exp\+\_\+params\+\_\+reset} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+md\+\_\+t $\ast$md\+\_\+p)]This function is attached to \hyperlink{group__fold__compound_structvrna__fc__s}{vrna\+\_\+fc\+\_\+s} objects as overloaded {\bfseries exp\+\_\+params\+\_\+reset()} method.

When no parameter is passed to this method, the resulting action is the same as passing {\itshape N\+U\+LL} as second parameter to \hyperlink{group__energy__parameters_gaa5409218068be84d7b50c78fbdaa85a9}{vrna\+\_\+exp\+\_\+params\+\_\+reset()}, i.\+e. global default model settings are used. Passing an object of type \hyperlink{group__model__details_structvrna__md__s}{vrna\+\_\+md\+\_\+s} resets the fold compound according to the specifications stored within the \hyperlink{group__model__details_structvrna__md__s}{vrna\+\_\+md\+\_\+s} object.  
\item[\label{wrappers__wrappers000067}%
\Hypertarget{wrappers__wrappers000067}%
Class \hyperlink{group__fold__compound}{vrna\+\_\+fc\+\_\+s} ]

This data structure is wrapped as an object {\bfseries fold\+\_\+compound} with several related functions attached as methods.

A new {\bfseries fold\+\_\+compound} can be obtained by calling one of its constructors\+:~\newline

\begin{DoxyItemize}
\item {\itshape fold\+\_\+compound(seq)} -- Initialize with a single sequence, or two concatenated sequences separated by an ampersand character \textquotesingle{}\&\textquotesingle{} (for cofolding)
\item {\itshape fold\+\_\+compound(aln)} -- Initialize with a sequence alignment {\itshape aln} stored as a list of sequences (with gap characters)
\end{DoxyItemize}The resulting object has a list of attached methods which in most cases directly correspond to functions that mainly operate on the corresponding {\itshape C} data structure\+:~\newline

\begin{DoxyItemize}
\item {\itshape \hyperlink{group__fold__compound_ac5eab693deac9a1a40c2a95ac294707c}{type()}} -- Get the type of the {\itshape fold\+\_\+compound} (See \hyperlink{group__fold__compound_ga01a4ff86fa71deaaa5d1abbd95a1447d}{vrna\+\_\+fc\+\_\+type\+\_\+e})
\item {\itshape \hyperlink{group__fold__compound_a95fbfed770b858e50c766505dc4bf998}{length()}} -- Get the length of the sequence(s) or alignment stored within the {\itshape fold\+\_\+compound} 
\end{DoxyItemize}
\item[\label{wrappers__wrappers000004}%
\Hypertarget{wrappers__wrappers000004}%
Global \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$filename, unsigned int options)]This function is attached as method {\bfseries file\+\_\+commands\+\_\+apply()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000059}%
\Hypertarget{wrappers__wrappers000059}%
Global \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format} (const char $\ast$filename, unsigned int options)]This function exists as an overloaded version where the {\ttfamily options} parameter may be omitted! In that case, the {\ttfamily options} parameter defaults to \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}.  
\item[\label{wrappers__wrappers000060}%
\Hypertarget{wrappers__wrappers000060}%
Global \hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read} (const char $\ast$filename, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)]In the target scripting language, only the first and last argument, {\ttfamily filename} and {\ttfamily options}, are passed to the corresponding function. The other arguments, which serve as output in the C-\/library, are available as additional return values. Hence, a function call in python may look like this\+:


\begin{DoxyCode}
num\_seq, names, aln, id, structure = RNA.file\_msa\_read(\textcolor{stringliteral}{"msa.stk"}, RNA.FILE\_FORMAT\_MSA\_STOCKHOLM)
\end{DoxyCode}


After successfully reading the first record, the variable {\ttfamily num\+\_\+seq} contains the number of sequences in the alignment (the actual return value of the C-\/function), while the variables {\ttfamily names}, {\ttfamily aln}, {\ttfamily id}, and {\ttfamily structure} are lists of the sequence names and aligned sequences, as well as strings holding the alignment ID and the structure as stated in the {\ttfamily S\+S\+\_\+cons} line, respectively. Note, the last two return values may be empty strings in case the alignment does not provide the required data.

This function exists as an overloaded version where the {\ttfamily options} parameter may be omitted! In that case, the {\ttfamily options} parameter defaults to \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}.  
\item[\label{wrappers__wrappers000061}%
\Hypertarget{wrappers__wrappers000061}%
Global \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record} (F\+I\+LE $\ast$fp, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)]In the target scripting language, only the first and last argument, {\ttfamily fp} and {\ttfamily options}, are passed to the corresponding function. The other arguments, which serve as output in the C-\/library, are available as additional return values. Hence, a function call in python may look like this\+:


\begin{DoxyCode}
f = open(\textcolor{stringliteral}{'msa.stk'}, \textcolor{charliteral}{'r'})
num\_seq, names, aln, \textcolor{keywordtype}{id}, structure = RNA.file\_msa\_read\_record(f, RNA.FILE\_FORMAT\_MSA\_STOCKHOLM)
f.close() 
\end{DoxyCode}


After successfully reading the first record, the variable {\ttfamily num\+\_\+seq} contains the number of sequences in the alignment (the actual return value of the C-\/function), while the variables {\ttfamily names}, {\ttfamily aln}, {\ttfamily id}, and {\ttfamily structure} are lists of the sequence names and aligned sequences, as well as strings holding the alignment ID and the structure as stated in the {\ttfamily S\+S\+\_\+cons} line, respectively. Note, the last two return values may be empty strings in case the alignment does not provide the required data.

This function exists as an overloaded version where the {\ttfamily options} parameter may be omitted! In that case, the {\ttfamily options} parameter defaults to \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}.  
\item[\label{wrappers__wrappers000062}%
\Hypertarget{wrappers__wrappers000062}%
Global \hyperlink{group__file__utils_ga0c90faf1515ce0b16620036b1055ff1e}{vrna\+\_\+file\+\_\+msa\+\_\+write} (const char $\ast$filename, const char $\ast$$\ast$names, const char $\ast$$\ast$aln, const char $\ast$id, const char $\ast$structure, const char $\ast$source, unsigned int options)]In the target scripting language, this function exists as a set of overloaded versions, where the last four parameters may be omitted. If the {\ttfamily options} parameter is missing the options default to (\hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM} $\vert$ \hyperlink{group__file__utils_ga1577ea0f497d9c501549c863a4f2c089}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND}).  
\item[\label{wrappers__wrappers000010}%
\Hypertarget{wrappers__wrappers000010}%
Global \hyperlink{group__hard__constraints_ga5b4de3247b67358080c176b94591a8e6}{vrna\+\_\+hc\+\_\+add\+\_\+from\+\_\+db} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$constraint, unsigned int options)]This function is attached as method {\bfseries hc\+\_\+add\+\_\+from\+\_\+db()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000006}%
\Hypertarget{wrappers__wrappers000006}%
Global \hyperlink{group__hard__constraints_ga36ff456c43bf920629cee5a236e4f0ff}{vrna\+\_\+hc\+\_\+init} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc)]This function is attached as method {\bfseries hc\+\_\+init()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000077}%
\Hypertarget{wrappers__wrappers000077}%
Class \hyperlink{group__model__details}{vrna\+\_\+md\+\_\+s} ]This data structure is wrapped as an object {\bfseries md} with multiple related functions attached as methods.

A new set of default parameters can be obtained by calling the constructure of {\bfseries md\+:} ~\newline

\begin{DoxyItemize}
\item {\itshape md()} -- Initialize with default settings
\end{DoxyItemize}

The resulting object has a list of attached methods which directly correspond to functions that mainly operate on the corresponding {\itshape C} data structure\+:~\newline

\begin{DoxyItemize}
\item {\itshape reset()} -- \hyperlink{group__model__details_ga8ac6ff84936282436f822644bf841f66}{vrna\+\_\+md\+\_\+set\+\_\+default()}
\item {\itshape set\+\_\+from\+\_\+globals()} -- \hyperlink{group__model__details_gabad896c3650d420f3f3ddefc69e2bceb}{set\+\_\+model\+\_\+details()}
\item {\itshape option\+\_\+string()} -- \hyperlink{group__model__details_ga3a7469f0725a849af6ba61a57dfd60ce}{vrna\+\_\+md\+\_\+option\+\_\+string()}
\end{DoxyItemize}

Note, that default parameters can be modified by directly setting any of the following global variables. Internally, getting/setting default parameters using their global variable representative translates into calls of the following functions, therefore these wrappers for these functions do not exist in the scripting language interface(s)\+:

\tabulinesep=1mm
\begin{longtabu} spread 0pt [c]{*{3}{|X[-1]}|}
\hline
\rowcolor{\tableheadbgcolor}\textbf{ global variable }&\textbf{ {\itshape C} getter }&\textbf{ {\itshape C} setter  }\\\cline{1-3}
\endfirsthead
\hline
\endfoot
\hline
\rowcolor{\tableheadbgcolor}\textbf{ global variable }&\textbf{ {\itshape C} getter }&\textbf{ {\itshape C} setter  }\\\cline{1-3}
\endhead
temperature &\hyperlink{group__model__details_ga96b24a74437f9ba46c4e06343155bf46}{vrna\+\_\+md\+\_\+defaults\+\_\+temperature\+\_\+get()} &\hyperlink{group__model__details_gaf9e527e9a2f7e6fd6e42bc6e602f5445}{vrna\+\_\+md\+\_\+defaults\+\_\+temperature()} \\\cline{1-3}
dangles &\hyperlink{group__model__details_ga67ca06f95ae133778c79a4493c9817b8}{vrna\+\_\+md\+\_\+defaults\+\_\+dangles\+\_\+get()} &\hyperlink{group__model__details_gac76a5374def8e5e4e644ff6e4cc72dee}{vrna\+\_\+md\+\_\+defaults\+\_\+dangles()} \\\cline{1-3}
beta\+Scale &\hyperlink{group__model__details_gabb8780f5410c52f970d75b044059bd09}{vrna\+\_\+md\+\_\+defaults\+\_\+beta\+Scale\+\_\+get()} &\hyperlink{group__model__details_gae984567db36c3f9b8731ecc917abf3a2}{vrna\+\_\+md\+\_\+defaults\+\_\+beta\+Scale()} \\\cline{1-3}
tetra\+\_\+loop &this is an alias of {\itshape special\+\_\+hp} &\\\cline{1-3}
special\+\_\+hp &\hyperlink{group__model__details_ga1d68a6efdaa1253cc63fd9cd06452559}{vrna\+\_\+md\+\_\+defaults\+\_\+special\+\_\+hp\+\_\+get()} &\hyperlink{group__model__details_gafff6449a02744add0308e653230c15fc}{vrna\+\_\+md\+\_\+defaults\+\_\+special\+\_\+hp()} \\\cline{1-3}
no\+Lonely\+Pairs &this is an alias of {\itshape no\+LP} &\\\cline{1-3}
no\+LP &\hyperlink{group__model__details_ga934344888fbacaed538bbbfe910f2aa6}{vrna\+\_\+md\+\_\+defaults\+\_\+no\+L\+P\+\_\+get()} &\hyperlink{group__model__details_ga2f88ffc393ac9d7987849c965fd29ea8}{vrna\+\_\+md\+\_\+defaults\+\_\+no\+L\+P()} \\\cline{1-3}
no\+GU &\hyperlink{group__model__details_ga5faa7d4e536d7fe36ec25428c0cf2563}{vrna\+\_\+md\+\_\+defaults\+\_\+no\+G\+U\+\_\+get()} &\hyperlink{group__model__details_ga98218f85c7a957a1d1ddf4627fdf5a39}{vrna\+\_\+md\+\_\+defaults\+\_\+no\+G\+U()} \\\cline{1-3}
no\+\_\+closing\+GU &this is an alias of {\itshape no\+G\+Uclosure} &\\\cline{1-3}
no\+G\+Uclosure &\hyperlink{group__model__details_ga4f7fdad083243a5348d63320ddaa70f3}{vrna\+\_\+md\+\_\+defaults\+\_\+no\+G\+Uclosure\+\_\+get()} &\hyperlink{group__model__details_gade5b9951d71ca2fb357a4e6c0c18ccd1}{vrna\+\_\+md\+\_\+defaults\+\_\+no\+G\+Uclosure()} \\\cline{1-3}
log\+ML &\hyperlink{group__model__details_ga93f04e070d529c5d0bb87c9681f6ad29}{vrna\+\_\+md\+\_\+defaults\+\_\+log\+M\+L\+\_\+get()} &\hyperlink{group__model__details_ga3de50a73455d88c3957386933b8e1f90}{vrna\+\_\+md\+\_\+defaults\+\_\+log\+M\+L()} \\\cline{1-3}
circ &\hyperlink{group__model__details_gad3a7e58de344ad93a08925f58f94f6fb}{vrna\+\_\+md\+\_\+defaults\+\_\+circ\+\_\+get()} &\hyperlink{group__model__details_ga4e1deb3e91a8a99e5c6dd905a5eb0186}{vrna\+\_\+md\+\_\+defaults\+\_\+circ()} \\\cline{1-3}
gquad &\hyperlink{group__model__details_gae645b8612f879eb38b45244fa9eddb9e}{vrna\+\_\+md\+\_\+defaults\+\_\+gquad\+\_\+get()} &\hyperlink{group__model__details_ga0685ca2aeb39af76f2421fc308163dce}{vrna\+\_\+md\+\_\+defaults\+\_\+gquad()} \\\cline{1-3}
uniq\+\_\+\+ML &\hyperlink{group__model__details_gab48e70fd024bf838404bcbcca0c874a0}{vrna\+\_\+md\+\_\+defaults\+\_\+uniq\+\_\+\+M\+L\+\_\+get()} &\hyperlink{group__model__details_ga59b944f61c5d2babec2d4c48c820de67}{vrna\+\_\+md\+\_\+defaults\+\_\+uniq\+\_\+\+M\+L()} \\\cline{1-3}
energy\+\_\+set &\hyperlink{group__model__details_ga017ed6afb1cba2b7f242412cab618b53}{vrna\+\_\+md\+\_\+defaults\+\_\+energy\+\_\+set\+\_\+get()} &\hyperlink{group__model__details_ga8dd29c55787a4576277e1907e92d810c}{vrna\+\_\+md\+\_\+defaults\+\_\+energy\+\_\+set()} \\\cline{1-3}
backtrack &\hyperlink{group__model__details_ga90da1156e6883ddd68527c2830706648}{vrna\+\_\+md\+\_\+defaults\+\_\+backtrack\+\_\+get()} &\hyperlink{group__model__details_ga978c468b2fe96a70d5191e3dd17d5599}{vrna\+\_\+md\+\_\+defaults\+\_\+backtrack()} \\\cline{1-3}
backtrack\+\_\+type &\hyperlink{group__model__details_ga1425b4ebd0e034dead66d79becd64143}{vrna\+\_\+md\+\_\+defaults\+\_\+backtrack\+\_\+type\+\_\+get()} &\hyperlink{group__model__details_ga68305274de96b56b7799575e222560d8}{vrna\+\_\+md\+\_\+defaults\+\_\+backtrack\+\_\+type()} \\\cline{1-3}
do\+\_\+backtrack &this is an alias of {\itshape compute\+\_\+bpp} &\\\cline{1-3}
compute\+\_\+bpp &\hyperlink{group__model__details_gaa3a537e61fbe0518673bf9f73fd820f3}{vrna\+\_\+md\+\_\+defaults\+\_\+compute\+\_\+bpp\+\_\+get()} &\hyperlink{group__model__details_gaf1b5db10f1f476767f9a95f8a78e3132}{vrna\+\_\+md\+\_\+defaults\+\_\+compute\+\_\+bpp()} \\\cline{1-3}
max\+\_\+bp\+\_\+span &\hyperlink{group__model__details_gaa60f989e062fecd4d4bac89c1883da85}{vrna\+\_\+md\+\_\+defaults\+\_\+max\+\_\+bp\+\_\+span\+\_\+get()} &\hyperlink{group__model__details_ga4c4bc962f09b4480cb8499f1cf8ae4ec}{vrna\+\_\+md\+\_\+defaults\+\_\+max\+\_\+bp\+\_\+span()} \\\cline{1-3}
min\+\_\+loop\+\_\+size &\hyperlink{group__model__details_ga5cc691174a75c652807dc361b617632a}{vrna\+\_\+md\+\_\+defaults\+\_\+min\+\_\+loop\+\_\+size\+\_\+get()} &\hyperlink{group__model__details_gac152f1e78c1058a10261022c8dfda0f7}{vrna\+\_\+md\+\_\+defaults\+\_\+min\+\_\+loop\+\_\+size()} \\\cline{1-3}
window\+\_\+size &\hyperlink{group__model__details_ga670146a9aa3ba77f4d422d60b7c30ac9}{vrna\+\_\+md\+\_\+defaults\+\_\+window\+\_\+size\+\_\+get()} &\hyperlink{group__model__details_ga7b802ce0e8c3181bf5cb580de6d5b26a}{vrna\+\_\+md\+\_\+defaults\+\_\+window\+\_\+size()} \\\cline{1-3}
old\+Ali\+En &\hyperlink{group__model__details_ga2374492b5019df88022fe4c05f0f3630}{vrna\+\_\+md\+\_\+defaults\+\_\+old\+Ali\+En\+\_\+get()} &\hyperlink{group__model__details_ga41521d5b9fb7e0f31e7ea73f5792afab}{vrna\+\_\+md\+\_\+defaults\+\_\+old\+Ali\+En()} \\\cline{1-3}
ribo &\hyperlink{group__model__details_ga169027f0c0561ea7d87b655e4b336bfc}{vrna\+\_\+md\+\_\+defaults\+\_\+ribo\+\_\+get()} &\hyperlink{group__model__details_ga937c45e1d06fd6168730a9b08d130be3}{vrna\+\_\+md\+\_\+defaults\+\_\+ribo()} \\\cline{1-3}
cv\+\_\+fact &\hyperlink{group__model__details_gae59c68393807217b0a2497adb64d3ee3}{vrna\+\_\+md\+\_\+defaults\+\_\+cv\+\_\+fact\+\_\+get()} &\hyperlink{group__model__details_gad3a3f40baafd91a6ce80a91a68e20053}{vrna\+\_\+md\+\_\+defaults\+\_\+cv\+\_\+fact()} \\\cline{1-3}
nc\+\_\+fact &\hyperlink{group__model__details_ga7ac759eaa7159bf5f022745f5da59508}{vrna\+\_\+md\+\_\+defaults\+\_\+nc\+\_\+fact\+\_\+get()} &\hyperlink{group__model__details_gac35e596c850dce3ad55c49119fd7d471}{vrna\+\_\+md\+\_\+defaults\+\_\+nc\+\_\+fact()} \\\cline{1-3}
sfact &\hyperlink{group__model__details_gab2df6aab954b63fd3592d18e90285dae}{vrna\+\_\+md\+\_\+defaults\+\_\+sfact\+\_\+get()} &\hyperlink{group__model__details_ga3f73d3029d3d0025d4cc311510cd95a3}{vrna\+\_\+md\+\_\+defaults\+\_\+sfact()} \\\cline{1-3}
\end{longtabu}

\item[\label{wrappers__wrappers000085}%
\Hypertarget{wrappers__wrappers000085}%
Global \hyperlink{group__pf__fold_gaa6b8983b559b9ef4b2e1b31113ea317b}{vrna\+\_\+mean\+\_\+bp\+\_\+distance} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc)]This function is attached as method {\bfseries \hyperlink{group__pf__fold_ga79cbc375af65f11609feb6b055269e7d}{mean\+\_\+bp\+\_\+distance()}} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000074}%
\Hypertarget{wrappers__wrappers000074}%
Global \hyperlink{group__mfe__fold_gabd3b147371ccf25c577f88bbbaf159fd}{vrna\+\_\+mfe} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, char $\ast$structure)]This function is attached as method {\bfseries mfe()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000075}%
\Hypertarget{wrappers__wrappers000075}%
Global \hyperlink{group__mfe__cofold_gaab22d10c1190f205f16a77cab9d5d3ee}{vrna\+\_\+mfe\+\_\+dimer} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, char $\ast$structure)]This function is attached as method {\bfseries mfe\+\_\+dimer()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000076}%
\Hypertarget{wrappers__wrappers000076}%
Global \hyperlink{group__local__mfe__fold_ga689df235a1915a1ad56e377383c044ce}{vrna\+\_\+mfe\+\_\+window} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, F\+I\+LE $\ast$file)]This function is attached as method {\bfseries mfe\+\_\+window()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000078}%
\Hypertarget{wrappers__wrappers000078}%
Global \hyperlink{group__neighbors_gab1473d856b15da2120fe648977798ad5}{vrna\+\_\+neighbors} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const short $\ast$pt, unsigned int options)]This function is attached as an overloaded method {\itshape neighbors()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily options} defaults to \hyperlink{group__neighbors_gaa5ffec4dd0d02df320f123e6888154d1}{V\+R\+N\+A\+\_\+\+M\+O\+V\+E\+S\+E\+T\+\_\+\+D\+E\+F\+A\+U\+LT} if it is omitted.  
\item[\label{wrappers__wrappers000081}%
\Hypertarget{wrappers__wrappers000081}%
Global \hyperlink{group__energy__parameters_gac40dc82e48a72a97cfc58b9da08a7792}{vrna\+\_\+params\+\_\+reset} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+md\+\_\+t $\ast$md\+\_\+p)]This function is attached to \hyperlink{group__fold__compound_structvrna__fc__s}{vrna\+\_\+fc\+\_\+s} objects as overloaded {\bfseries params\+\_\+reset()} method.

When no parameter is passed to this method, the resulting action is the same as passing {\itshape N\+U\+LL} as second parameter to \hyperlink{group__energy__parameters_gac40dc82e48a72a97cfc58b9da08a7792}{vrna\+\_\+params\+\_\+reset()}, i.\+e. global default model settings are used. Passing an object of type \hyperlink{group__model__details_structvrna__md__s}{vrna\+\_\+md\+\_\+s} resets the fold compound according to the specifications stored within the \hyperlink{group__model__details_structvrna__md__s}{vrna\+\_\+md\+\_\+s} object.  
\item[\label{wrappers__wrappers000079}%
\Hypertarget{wrappers__wrappers000079}%
Global \hyperlink{group__energy__parameters_ga5d1909208f7ea3baa98b75afaa1f62ca}{vrna\+\_\+params\+\_\+subst} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+param\+\_\+t $\ast$par)]This function is attached to \hyperlink{group__fold__compound_structvrna__fc__s}{vrna\+\_\+fc\+\_\+s} objects as {\bfseries params\+\_\+subst()} method.  
\item[\label{wrappers__wrappers000093}%
\Hypertarget{wrappers__wrappers000093}%
Global \hyperlink{group__paths_gab6aee4143f8b103518d5cbfe6bfe5eae}{vrna\+\_\+path} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, short $\ast$pt, unsigned int steps, unsigned int options)]This function is attached as an overloaded method {\itshape path()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily options} defaults to \hyperlink{group__paths_gaf1bb1f8fec8d24e3b821f621b19f77b4}{V\+R\+N\+A\+\_\+\+P\+A\+T\+H\+\_\+\+D\+E\+F\+A\+U\+LT} if it is omitted.  
\item[\label{wrappers__wrappers000065}%
\Hypertarget{wrappers__wrappers000065}%
Global \hyperlink{group__direct__paths_ga4b2283c4142cafd99678495585fcc842}{vrna\+\_\+path\+\_\+findpath} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$s1, const char $\ast$s2, int width)]This function is attached as an overloaded method {\itshape path\+\_\+findpath()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily width} defaults to 1 if it is omitted.  
\item[\label{wrappers__wrappers000063}%
\Hypertarget{wrappers__wrappers000063}%
Global \hyperlink{group__direct__paths_gad611574a76593e26021f177e7854b6b4}{vrna\+\_\+path\+\_\+findpath\+\_\+saddle} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$s1, const char $\ast$s2, int width)]This function is attached as an overloaded method {\itshape path\+\_\+findpath\+\_\+saddle()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily width} defaults to 1 if it is omitted.  
\item[\label{wrappers__wrappers000064}%
\Hypertarget{wrappers__wrappers000064}%
Global \hyperlink{group__direct__paths_gada8d722e37401b1aea30128b07555771}{vrna\+\_\+path\+\_\+findpath\+\_\+saddle\+\_\+ub} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$s1, const char $\ast$s2, int width, int maxE)]This function is attached as an overloaded method {\itshape path\+\_\+findpath\+\_\+saddle()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily width} defaults to 1 if it is omitted, while the optional parameter {\ttfamily maxE} defaults to \hyperlink{energy__const_8h_a12c2040f25d8e3a7b9e1c2024c618cb6}{I\+NF}. In case the function did not find a path with $E_{saddle} < E_{max}$ the function returns a {\itshape N\+U\+LL} object, i.\+e. {\itshape undef} for Perl and {\itshape None} for Python.  
\item[\label{wrappers__wrappers000066}%
\Hypertarget{wrappers__wrappers000066}%
Global \hyperlink{group__direct__paths_ga8d47812616303f40057dfb033869863a}{vrna\+\_\+path\+\_\+findpath\+\_\+ub} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$s1, const char $\ast$s2, int width, int maxE)]This function is attached as an overloaded method {\itshape path\+\_\+findpath()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily width} defaults to 1 if it is omitted, while the optional parameter {\ttfamily maxE} defaults to \hyperlink{energy__const_8h_a12c2040f25d8e3a7b9e1c2024c618cb6}{I\+NF}. In case the function did not find a path with $E_{saddle} < E_{max}$ the function returns an empty list.  
\item[\label{wrappers__wrappers000094}%
\Hypertarget{wrappers__wrappers000094}%
Global \hyperlink{group__paths_gae92cce443a8a64f7b7fb89867b7d6125}{vrna\+\_\+path\+\_\+gradient} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, short $\ast$pt, unsigned int options)]This function is attached as an overloaded method {\itshape path\+\_\+gradient()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily options} defaults to \hyperlink{group__paths_gaf1bb1f8fec8d24e3b821f621b19f77b4}{V\+R\+N\+A\+\_\+\+P\+A\+T\+H\+\_\+\+D\+E\+F\+A\+U\+LT} if it is omitted.  
\item[\label{wrappers__wrappers000095}%
\Hypertarget{wrappers__wrappers000095}%
Global \hyperlink{group__paths_ga9234756c337078fa599529d3db70d913}{vrna\+\_\+path\+\_\+random} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, short $\ast$pt, unsigned int steps, unsigned int options)]This function is attached as an overloaded method {\itshape path\+\_\+gradient()} to objects of type {\itshape fold\+\_\+compound}. The optional parameter {\ttfamily options} defaults to \hyperlink{group__paths_gaf1bb1f8fec8d24e3b821f621b19f77b4}{V\+R\+N\+A\+\_\+\+P\+A\+T\+H\+\_\+\+D\+E\+F\+A\+U\+LT} if it is omitted.  
\item[\label{wrappers__wrappers000087}%
\Hypertarget{wrappers__wrappers000087}%
Global \hyperlink{group__subopt__stochbt_ga901fe42a33b07be083421741bf7dc610}{vrna\+\_\+pbacktrack} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc)]This function is attached as overloaded method {\bfseries \hyperlink{group__subopt__stochbt_gac03ca6db186bb3bf0a2a326d7fb3ba03}{pbacktrack()}} to objects of type {\itshape fold\+\_\+compound} that accepts an optional {\ttfamily length} argument. Hence, it serves as a replacement for \hyperlink{group__subopt__stochbt_ga901fe42a33b07be083421741bf7dc610}{vrna\+\_\+pbacktrack()}.  
\item[\label{wrappers__wrappers000086}%
\Hypertarget{wrappers__wrappers000086}%
Global \hyperlink{group__subopt__stochbt_ga347375a4da8bdff74639529847f97d4f}{vrna\+\_\+pbacktrack5} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int length)]This function is attached as overloaded method {\bfseries \hyperlink{group__subopt__stochbt_gac03ca6db186bb3bf0a2a326d7fb3ba03}{pbacktrack()}} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000083}%
\Hypertarget{wrappers__wrappers000083}%
Global \hyperlink{group__pf__fold_ga29e256d688ad221b78d37f427e0e99bc}{vrna\+\_\+pf} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, char $\ast$structure)]This function is attached as method {\bfseries pf()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000084}%
\Hypertarget{wrappers__wrappers000084}%
Global \hyperlink{group__pf__cofold_ga4e5c7d06c302a7c59fc0d64dc142ca63}{vrna\+\_\+pf\+\_\+dimer} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, char $\ast$structure)]This function is attached as method {\bfseries pf\+\_\+dimer()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000018}%
\Hypertarget{wrappers__wrappers000018}%
Global \hyperlink{group__soft__constraints_gaf162aedac7422f2eb16ea030f47d2f4b}{vrna\+\_\+sc\+\_\+add\+\_\+bp} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, int j, F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+BL energy, unsigned int options)]

This function is attached as an overloaded method {\bfseries sc\+\_\+add\+\_\+bp()} to objects of type {\itshape fold\+\_\+compound}. The method either takes arguments for a single base pair (i,j) with the corresponding energy value\+: 
\begin{DoxyCode}
fold\_compound.sc\_add\_bp(i, j, energy, options)
\end{DoxyCode}
 or an entire 2-\/dimensional matrix with dimensions n x n that stores free energy contributions for any base pair (i,j) with $ 1 \leq i < j \leq n $\+: 
\begin{DoxyCode}
fold\_compound.sc\_add\_bp(matrix, options)
\end{DoxyCode}
 In both variants, the {\ttfamily options} argument is optional can may be omitted.  
\item[\label{wrappers__wrappers000023}%
\Hypertarget{wrappers__wrappers000023}%
Global \hyperlink{group__soft__constraints_gabde7d07a79bb9a8f4721aee247b674ea}{vrna\+\_\+sc\+\_\+add\+\_\+bt} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+callback\+\_\+sc\+\_\+backtrack $\ast$f)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+bt()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000021}%
\Hypertarget{wrappers__wrappers000021}%
Global \hyperlink{group__soft__constraints_ga15c6d52471ec97897e2bb7f964f5deb6}{vrna\+\_\+sc\+\_\+add\+\_\+data} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, void $\ast$data, vrna\+\_\+callback\+\_\+free\+\_\+auxdata $\ast$free\+\_\+data)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+data()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000024}%
\Hypertarget{wrappers__wrappers000024}%
Global \hyperlink{group__soft__constraints_ga87e382b5d0c9b7d9ce1b79c0473ff700}{vrna\+\_\+sc\+\_\+add\+\_\+exp\+\_\+f} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+callback\+\_\+sc\+\_\+exp\+\_\+energy $\ast$exp\+\_\+f)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+exp\+\_\+f()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000022}%
\Hypertarget{wrappers__wrappers000022}%
Global \hyperlink{group__soft__constraints_ga8c7d907ec0125cd61c04e0908010a4e9}{vrna\+\_\+sc\+\_\+add\+\_\+f} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+callback\+\_\+sc\+\_\+energy $\ast$f)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+f()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000011}%
\Hypertarget{wrappers__wrappers000011}%
Global \hyperlink{group__constraints__ligand_gaa6ff0113a3a76dc0b8d62961f4e1dfa0}{vrna\+\_\+sc\+\_\+add\+\_\+hi\+\_\+motif} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$seq, const char $\ast$structure, F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+BL energy, unsigned int options)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+hi\+\_\+motif()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000012}%
\Hypertarget{wrappers__wrappers000012}%
Global \hyperlink{group__SHAPE__reactivities_ga57d612b58e1c61dd6cfcb5a843f8f1b3}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const double $\ast$reactivities, double m, double b, unsigned int options)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000013}%
\Hypertarget{wrappers__wrappers000013}%
Global \hyperlink{group__SHAPE__reactivities_ga04ba85da63d8c793bb8001d1e6f800ba}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$$\ast$shape\+\_\+files, const int $\ast$shape\+\_\+file\+\_\+association, double m, double b, unsigned int options)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+deigan\+\_\+ali()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000014}%
\Hypertarget{wrappers__wrappers000014}%
Global \hyperlink{group__SHAPE__reactivities_gaf3c65a045060aef5c4e41693d30af58c}{vrna\+\_\+sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const double $\ast$reactivities, double b, double default\+\_\+value, const char $\ast$shape\+\_\+conversion, unsigned int options)]This function is attached as method {\bfseries sc\+\_\+add\+\_\+\+S\+H\+A\+P\+E\+\_\+zarringhalam()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000020}%
\Hypertarget{wrappers__wrappers000020}%
Global \hyperlink{group__soft__constraints_ga069915fe203a2c8e522dd37847177a09}{vrna\+\_\+sc\+\_\+add\+\_\+up} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int i, F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+BL energy, unsigned int options)]This function is attached as an overloaded method {\bfseries sc\+\_\+add\+\_\+up()} to objects of type {\itshape fold\+\_\+compound}. The method either takes arguments for a single nucleotide $i $ with the corresponding energy value\+: 
\begin{DoxyCode}
fold\_compound.sc\_add\_up(i, energy, options)
\end{DoxyCode}
 or an entire vector that stores free energy contributions for each nucleotide $i $ with $ 1 \leq i \leq n $\+: 
\begin{DoxyCode}
fold\_compound.sc\_add\_bp(vector, options)
\end{DoxyCode}
 In both variants, the {\ttfamily options} argument is optional can may be omitted.  
\item[\label{wrappers__wrappers000015}%
\Hypertarget{wrappers__wrappers000015}%
Global \hyperlink{group__soft__constraints_ga9d977a1681356778cc66dceafbe5b032}{vrna\+\_\+sc\+\_\+init} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc)]This function is attached as method {\bfseries sc\+\_\+init()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000016}%
\Hypertarget{wrappers__wrappers000016}%
Global \hyperlink{group__soft__constraints_ga73cdc07b9a199c614367bebef0f2c41a}{vrna\+\_\+sc\+\_\+remove} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc)]This function is attached as method {\bfseries sc\+\_\+remove()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000017}%
\Hypertarget{wrappers__wrappers000017}%
Global \hyperlink{group__soft__constraints_ga8e4334b24bc91453fbcda490a4e331af}{vrna\+\_\+sc\+\_\+set\+\_\+bp} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+BL $\ast$$\ast$constraints, unsigned int options)]This function is attached as method {\bfseries sc\+\_\+set\+\_\+bp()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000019}%
\Hypertarget{wrappers__wrappers000019}%
Global \hyperlink{group__soft__constraints_ga99ed63f3ef9e7fe3997932030487a344}{vrna\+\_\+sc\+\_\+set\+\_\+up} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const F\+L\+T\+\_\+\+O\+R\+\_\+\+D\+BL $\ast$constraints, unsigned int options)]This function is attached as method {\bfseries sc\+\_\+set\+\_\+up()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000090}%
\Hypertarget{wrappers__wrappers000090}%
Global \hyperlink{group__subopt__wuchty_ga0f11d738fb8c8b1885a90c11c8931ff6}{vrna\+\_\+subopt} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int delta, int sorted, F\+I\+LE $\ast$fp)]This function is attached as method {\bfseries \hyperlink{group__subopt__wuchty_ga700f662506a233e42dd7fda74fafd40e}{subopt()}} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000091}%
\Hypertarget{wrappers__wrappers000091}%
Global \hyperlink{group__subopt__wuchty_ga1053837e6b6f158093508f8a70998352}{vrna\+\_\+subopt\+\_\+cb} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, int delta, vrna\+\_\+subopt\+\_\+callback $\ast$cb, void $\ast$data)]This function is attached as method {\bfseries subopt\+\_\+cb()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000092}%
\Hypertarget{wrappers__wrappers000092}%
Global \hyperlink{group__subopt__zuker_gababde9d210eb433854f1e71da7815fbb}{vrna\+\_\+subopt\+\_\+zuker} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc)]This function is attached as method {\bfseries subopt\+\_\+zuker()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000068}%
\Hypertarget{wrappers__wrappers000068}%
Global \hyperlink{group__domains__up_gaec0c3313fb2951946614f920d289829a}{vrna\+\_\+ud\+\_\+add\+\_\+motif} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, const char $\ast$motif, double motif\+\_\+en, unsigned int loop\+\_\+type)]This function is attached as method {\bfseries ud\+\_\+add\+\_\+motif()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000069}%
\Hypertarget{wrappers__wrappers000069}%
Global \hyperlink{group__domains__up_gada59cb0c498b812eadd010811af3f2d4}{vrna\+\_\+ud\+\_\+remove} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc)]This function is attached as method {\bfseries ud\+\_\+remove()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000070}%
\Hypertarget{wrappers__wrappers000070}%
Global \hyperlink{group__domains__up_gac1f18c312b91d80089534a87d956e58b}{vrna\+\_\+ud\+\_\+set\+\_\+data} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, void $\ast$data, vrna\+\_\+callback\+\_\+free\+\_\+auxdata $\ast$free\+\_\+cb)]This function is attached as method {\bfseries ud\+\_\+set\+\_\+data()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000072}%
\Hypertarget{wrappers__wrappers000072}%
Global \hyperlink{group__domains__up_ga2fb1db2099da26c76247e1209ad4aa09}{vrna\+\_\+ud\+\_\+set\+\_\+exp\+\_\+prod\+\_\+rule\+\_\+cb} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+callback\+\_\+ud\+\_\+exp\+\_\+production $\ast$pre\+\_\+cb, vrna\+\_\+callback\+\_\+ud\+\_\+exp\+\_\+energy $\ast$exp\+\_\+e\+\_\+cb)]This function is attached as method {\bfseries ud\+\_\+set\+\_\+exp\+\_\+prod\+\_\+rule\+\_\+cb()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000073}%
\Hypertarget{wrappers__wrappers000073}%
Global \hyperlink{unstructured__domains_8h_a13ac877c9db89a1a5b5d9c0394148595}{vrna\+\_\+ud\+\_\+set\+\_\+prob\+\_\+cb} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+callback\+\_\+ud\+\_\+probs\+\_\+add $\ast$setter, vrna\+\_\+callback\+\_\+ud\+\_\+probs\+\_\+get $\ast$getter)]This function is attached as method {\bfseries ud\+\_\+set\+\_\+prob\+\_\+cb()} to objects of type {\itshape fold\+\_\+compound}  
\item[\label{wrappers__wrappers000071}%
\Hypertarget{wrappers__wrappers000071}%
Global \hyperlink{group__domains__up_ga745a99f0bc72898d54de16f6e538828a}{vrna\+\_\+ud\+\_\+set\+\_\+prod\+\_\+rule\+\_\+cb} (vrna\+\_\+fold\+\_\+compound\+\_\+t $\ast$vc, vrna\+\_\+callback\+\_\+ud\+\_\+production $\ast$pre\+\_\+cb, vrna\+\_\+callback\+\_\+ud\+\_\+energy $\ast$e\+\_\+cb)]This function is attached as method {\bfseries ud\+\_\+set\+\_\+prod\+\_\+rule\+\_\+cb()} to objects of type {\itshape fold\+\_\+compound} 
\end{DoxyRefList}