\hypertarget{group__file__utils}{}\section{Functions to Read/\+Write several File Formats for R\+NA Sequences, Structures, and Alignments}
\label{group__file__utils}\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}


Functions dealing with file formats for R\+NA sequences, structures, and alignments.  




\subsection{Detailed Description}
Functions dealing with file formats for R\+NA sequences, structures, and alignments. 

Collaboration diagram for Functions to Read/\+Write several File Formats for R\+NA Sequences, Structures, and Alignments\+:
\nopagebreak
\begin{figure}[H]
\begin{center}
\leavevmode
\includegraphics[width=327pt]{group__file__utils}
\end{center}
\end{figure}
\subsection*{Files}
\begin{DoxyCompactItemize}
\item 
file \hyperlink{commands_8h}{commands.\+h}
\begin{DoxyCompactList}\small\item\em Parse and apply different commands that alter the behavior of secondary structure prediction and evaluation. \end{DoxyCompactList}\item 
file \hyperlink{file__formats_8h}{file\+\_\+formats.\+h}
\begin{DoxyCompactList}\small\item\em Read and write different file formats for R\+NA sequences, structures. \end{DoxyCompactList}\item 
file \hyperlink{file__formats__msa_8h}{file\+\_\+formats\+\_\+msa.\+h}
\begin{DoxyCompactList}\small\item\em Functions dealing with file formats for Multiple Sequence Alignments (M\+SA) \end{DoxyCompactList}\item 
file \hyperlink{file__utils_8h}{file\+\_\+utils.\+h}
\begin{DoxyCompactList}\small\item\em Several utilities for file handling. \end{DoxyCompactList}\item 
file \hyperlink{ribo_8h}{ribo.\+h}
\begin{DoxyCompactList}\small\item\em Parse Ribo\+Sum Scoring Matrices for Covariance Scoring of Alignments. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Data Structures}
\begin{DoxyCompactItemize}
\item 
struct \hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s}
\begin{DoxyCompactList}\small\item\em List element for commands ready for application to a \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t}.  \hyperlink{group__file__utils_structvrna__command__s}{More...}\end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Macros}
\begin{DoxyCompactItemize}
\item 
\#define \hyperlink{group__file__utils_gac54dec838d7b6bebd5df85f71702d324}{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+HC}~1U
\begin{DoxyCompactList}\small\item\em Command parse/apply flag indicating hard constraints. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga8cad3c1f83e6f149829c49a186a83e21}{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SC}~2U
\begin{DoxyCompactList}\small\item\em Command parse/apply flag indicating soft constraints. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga6c6409780698826b04ebfed9151d7649}{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+UD}~4U
\begin{DoxyCompactList}\small\item\em Command parse/apply flag indicating unstructured domains. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gaf5e20210173cdb83bf70256a454f284b}{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SD}~8U
\begin{DoxyCompactList}\small\item\em Command parse/apply flag indicating structured domains. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga0a6c88e21e366dca14958d69cd024008}{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+D\+E\+F\+A\+U\+L\+TS}
\begin{DoxyCompactList}\small\item\em Command parse/apply flag indicating default set of commands. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gabec89c09874528c6cb73140a4c3d86d7}{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}~32U
\begin{DoxyCompactList}\small\item\em Tell a function that an input is assumed to span several lines. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga7d725ef525b29891abef3f1ed42599a4}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}~32U
\begin{DoxyCompactList}\small\item\em parse multiline constraint \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}~1U
\begin{DoxyCompactList}\small\item\em Option flag indicating ClustalW formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}~2U
\begin{DoxyCompactList}\small\item\em Option flag indicating Stockholm 1.\+0 formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}~4U
\begin{DoxyCompactList}\small\item\em Option flag indicating F\+A\+S\+TA (Pearson) formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}~8U
\begin{DoxyCompactList}\small\item\em Option flag indicating M\+AF formatted files. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga494488a771aa0c602fb4cf445be34d47}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS}~16U
\begin{DoxyCompactList}\small\item\em Option flag indicating most informative sequence (M\+IS) output. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}
\begin{DoxyCompactList}\small\item\em Option flag indicating the set of default file formats. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}~4096U
\begin{DoxyCompactList}\small\item\em Option flag to disable validation of the alignment. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gabdc948f547e550125de3e7c65878400c}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}~8192U
\begin{DoxyCompactList}\small\item\em Return flag of \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} to indicate unknown or malformatted alignment. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga1577ea0f497d9c501549c863a4f2c089}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND}~16384U
\begin{DoxyCompactList}\small\item\em Option flag indicating to append data to a multiple sequence alignment file rather than overwriting it. \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_ga73aa3701a22c5acfcb2e8927287a0554}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+Q\+U\+I\+ET}~32768U
\begin{DoxyCompactList}\small\item\em Option flag to suppress unnecessary spam messages on {\ttfamily stderr} \end{DoxyCompactList}\item 
\#define \hyperlink{group__file__utils_gad98ca08968d19f221f29d2a1faf26887}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+I\+L\+E\+NT}~65536U
\begin{DoxyCompactList}\small\item\em Option flag to completely silence any warnings on {\ttfamily stderr} \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Typedefs}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}\label{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}} 
typedef struct \hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s} \hyperlink{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}{vrna\+\_\+cmd\+\_\+t}
\begin{DoxyCompactList}\small\item\em Typename for the command repesenting data structure \hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s}. \end{DoxyCompactList}\end{DoxyCompactItemize}
\subsection*{Enumerations}
\begin{DoxyCompactItemize}
\item 
\mbox{\Hypertarget{group__file__utils_ga2ea3e452bf3b3f3ca513b3b081a86137}\label{group__file__utils_ga2ea3e452bf3b3f3ca513b3b081a86137}} 
enum \hyperlink{group__file__utils_ga2ea3e452bf3b3f3ca513b3b081a86137}{vrna\+\_\+command\+\_\+e} \begin{DoxyCompactList}\small\item\em Types of commands within a list of \hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s} structures. \end{DoxyCompactList}
\end{DoxyCompactItemize}
\subsection*{Functions}
\begin{DoxyCompactItemize}
\item 
\hyperlink{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}{vrna\+\_\+cmd\+\_\+t} $\ast$ \hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\+\_\+file\+\_\+commands\+\_\+read} (const char $\ast$filename, unsigned int options)
\begin{DoxyCompactList}\small\item\em Extract a list of commands from a command file. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply} (\hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$vc, const char $\ast$filename, unsigned int options)
\begin{DoxyCompactList}\small\item\em Apply a list of commands from a command file. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply} (\hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$vc, \hyperlink{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}{vrna\+\_\+cmd\+\_\+t} $\ast$commands, unsigned int options)
\begin{DoxyCompactList}\small\item\em Apply a list of commands to a \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t}. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_ga926fd7632e68f3da2e2069402b80b1ab}{vrna\+\_\+commands\+\_\+free} (\hyperlink{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}{vrna\+\_\+cmd\+\_\+t} $\ast$commands)
\begin{DoxyCompactList}\small\item\em Free memory occupied by a list of commands. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_gaaface7db12fadc3d271641c4515ab6e4}{vrna\+\_\+file\+\_\+helixlist} (const char $\ast$seq, const char $\ast$db, float energy, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure as helix list. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_gab69682373ccca1e0e28cc967eec07745}{vrna\+\_\+file\+\_\+connect} (const char $\ast$seq, const char $\ast$db, float energy, const char $\ast$identifier, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure as connect table. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_ga9b462e6f202594af5d3fa56e280d633f}{vrna\+\_\+file\+\_\+bpseq} (const char $\ast$seq, const char $\ast$db, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure in bpseq format. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_ga31f4a6c2ea1495a6e4f9eb45a9f6193d}{vrna\+\_\+file\+\_\+json} (const char $\ast$seq, const char $\ast$db, double energy, const char $\ast$identifier, F\+I\+LE $\ast$file)
\begin{DoxyCompactList}\small\item\em Print a secondary structure in jsonformat. \end{DoxyCompactList}\item 
unsigned int \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record} (char $\ast$$\ast$header, char $\ast$$\ast$sequence, char $\ast$$\ast$$\ast$rest, F\+I\+LE $\ast$file, unsigned int options)
\begin{DoxyCompactList}\small\item\em Get a (fasta) data set from a file or stdin. \end{DoxyCompactList}\item 
char $\ast$ \hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure} (const char $\ast$$\ast$lines, unsigned int length, unsigned int option)
\begin{DoxyCompactList}\small\item\em Extract a dot-\/bracket structure string from (multiline)character array. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_ga646ebf45450a69a7f2533f9ecd283a32}{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read} (const char $\ast$file\+\_\+name, int length, double default\+\_\+value, char $\ast$sequence, double $\ast$values)
\begin{DoxyCompactList}\small\item\em Read data from a given S\+H\+A\+PE reactivity input file. \end{DoxyCompactList}\item 
void \hyperlink{group__file__utils_ga55a9ae6dfeecc1b3f0c2acf6fa796c15}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint} (char $\ast$$\ast$cstruc, const char $\ast$$\ast$lines, unsigned int option)
\begin{DoxyCompactList}\small\item\em Extract a hard constraint encoded as pseudo dot-\/bracket string. \end{DoxyCompactList}\item 
unsigned int \hyperlink{group__file__utils_gafd194a69af9d92b5b0412a7627ac1595}{read\+\_\+record} (char $\ast$$\ast$header, char $\ast$$\ast$sequence, char $\ast$$\ast$$\ast$rest, unsigned int options)
\begin{DoxyCompactList}\small\item\em Get a data record from stdin. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read} (const char $\ast$filename, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)
\begin{DoxyCompactList}\small\item\em Read a multiple sequence alignment from file. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record} (F\+I\+LE $\ast$fp, char $\ast$$\ast$$\ast$names, char $\ast$$\ast$$\ast$aln, char $\ast$$\ast$id, char $\ast$$\ast$structure, unsigned int options)
\begin{DoxyCompactList}\small\item\em Read a multiple sequence alignment from file handle. \end{DoxyCompactList}\item 
unsigned int \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format} (const char $\ast$filename, unsigned int options)
\begin{DoxyCompactList}\small\item\em Detect the format of a multiple sequence alignment file. \end{DoxyCompactList}\item 
int \hyperlink{group__file__utils_ga0c90faf1515ce0b16620036b1055ff1e}{vrna\+\_\+file\+\_\+msa\+\_\+write} (const char $\ast$filename, const char $\ast$$\ast$names, const char $\ast$$\ast$aln, const char $\ast$id, const char $\ast$structure, const char $\ast$source, unsigned int options)
\begin{DoxyCompactList}\small\item\em Write multiple sequence alignment file. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{group__file__utils_ga4382a56d2fee9ed738364b99329edc7c}\label{group__file__utils_ga4382a56d2fee9ed738364b99329edc7c}} 
void \hyperlink{group__file__utils_ga4382a56d2fee9ed738364b99329edc7c}{vrna\+\_\+file\+\_\+copy} (F\+I\+LE $\ast$from, F\+I\+LE $\ast$to)
\begin{DoxyCompactList}\small\item\em Inefficient `cp\textquotesingle{}. \end{DoxyCompactList}\item 
char $\ast$ \hyperlink{group__file__utils_gad16c270b70a77c753088b29de8a802dc}{vrna\+\_\+read\+\_\+line} (F\+I\+LE $\ast$fp)
\begin{DoxyCompactList}\small\item\em Read a line of arbitrary length from a stream. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{group__file__utils_gacfa44a974272a4d42c5d391cc780a0ec}\label{group__file__utils_gacfa44a974272a4d42c5d391cc780a0ec}} 
int \hyperlink{group__file__utils_gacfa44a974272a4d42c5d391cc780a0ec}{vrna\+\_\+mkdir\+\_\+p} (const char $\ast$path)
\begin{DoxyCompactList}\small\item\em Recursivly create a directory tree. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{group__file__utils_ga8d0f8f84f1e2628172cf3c8648b2e507}\label{group__file__utils_ga8d0f8f84f1e2628172cf3c8648b2e507}} 
char $\ast$ \hyperlink{group__file__utils_ga8d0f8f84f1e2628172cf3c8648b2e507}{vrna\+\_\+basename} (const char $\ast$path)
\begin{DoxyCompactList}\small\item\em Extract the filename from a file path. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{group__file__utils_gae006d31b088d52d8aace86948def5fe8}\label{group__file__utils_gae006d31b088d52d8aace86948def5fe8}} 
char $\ast$ \hyperlink{group__file__utils_gae006d31b088d52d8aace86948def5fe8}{vrna\+\_\+dirname} (const char $\ast$path)
\begin{DoxyCompactList}\small\item\em Extract the directory part of a file path. \end{DoxyCompactList}\item 
char $\ast$ \hyperlink{group__file__utils_ga72eae0a472bad905b6facccf5b453a76}{vrna\+\_\+filename\+\_\+sanitize} (const char $\ast$name, const char $\ast$replacement)
\begin{DoxyCompactList}\small\item\em Sanitize a file name. \end{DoxyCompactList}\item 
\mbox{\Hypertarget{group__file__utils_ga5e125c9586fcd4e2e1559fe76f7289cc}\label{group__file__utils_ga5e125c9586fcd4e2e1559fe76f7289cc}} 
float $\ast$$\ast$ \hyperlink{group__file__utils_ga5e125c9586fcd4e2e1559fe76f7289cc}{readribosum} (char $\ast$name)
\begin{DoxyCompactList}\small\item\em Read a Ribo\+Sum or other user-\/defined Scoring Matrix and Store into global Memory. \end{DoxyCompactList}\end{DoxyCompactItemize}


\subsection{Data Structure Documentation}
\index{vrna\+\_\+command\+\_\+s@{vrna\+\_\+command\+\_\+s}}\label{structvrna__command__s}
\Hypertarget{group__file__utils_structvrna__command__s}
\subsubsection{struct vrna\+\_\+command\+\_\+s}
List element for commands ready for application to a \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t}. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\+\_\+file\+\_\+commands\+\_\+read()}, \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_ga926fd7632e68f3da2e2069402b80b1ab}{vrna\+\_\+commands\+\_\+free()} 
\end{DoxySeeAlso}


\subsection{Macro Definition Documentation}
\mbox{\Hypertarget{group__file__utils_gac54dec838d7b6bebd5df85f71702d324}\label{group__file__utils_gac54dec838d7b6bebd5df85f71702d324}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+HC@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+HC}}
\index{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+HC@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+HC}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+HC}{VRNA\_CMD\_PARSE\_HC}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+HC~1U}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Command parse/apply flag indicating hard constraints. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s}, \hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\+\_\+file\+\_\+commands\+\_\+read()}, \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga8cad3c1f83e6f149829c49a186a83e21}\label{group__file__utils_ga8cad3c1f83e6f149829c49a186a83e21}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SC@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SC}}
\index{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SC@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SC}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SC}{VRNA\_CMD\_PARSE\_SC}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SC~2U}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Command parse/apply flag indicating soft constraints. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s}, \hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\+\_\+file\+\_\+commands\+\_\+read()}, \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga6c6409780698826b04ebfed9151d7649}\label{group__file__utils_ga6c6409780698826b04ebfed9151d7649}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+UD@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+UD}}
\index{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+UD@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+UD}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+UD}{VRNA\_CMD\_PARSE\_UD}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+UD~4U}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Command parse/apply flag indicating unstructured domains. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s}, \hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\+\_\+file\+\_\+commands\+\_\+read()}, \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_gaf5e20210173cdb83bf70256a454f284b}\label{group__file__utils_gaf5e20210173cdb83bf70256a454f284b}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SD@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SD}}
\index{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SD@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SD}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SD}{VRNA\_CMD\_PARSE\_SD}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+SD~8U}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Command parse/apply flag indicating structured domains. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s}, \hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\+\_\+file\+\_\+commands\+\_\+read()}, \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga0a6c88e21e366dca14958d69cd024008}\label{group__file__utils_ga0a6c88e21e366dca14958d69cd024008}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+D\+E\+F\+A\+U\+L\+TS@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+D\+E\+F\+A\+U\+L\+TS}}
\index{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+D\+E\+F\+A\+U\+L\+TS@{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+D\+E\+F\+A\+U\+L\+TS}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+D\+E\+F\+A\+U\+L\+TS}{VRNA\_CMD\_PARSE\_DEFAULTS}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+C\+M\+D\+\_\+\+P\+A\+R\+S\+E\+\_\+\+D\+E\+F\+A\+U\+L\+TS}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}

{\bfseries Value\+:}
\begin{DoxyCode}
(\hyperlink{group__file__utils_gac54dec838d7b6bebd5df85f71702d324}{VRNA\_CMD\_PARSE\_HC} \(\backslash\)
                                 | \hyperlink{group__file__utils_ga8cad3c1f83e6f149829c49a186a83e21}{VRNA\_CMD\_PARSE\_SC} \(\backslash\)
                                 | \hyperlink{group__file__utils_ga6c6409780698826b04ebfed9151d7649}{VRNA\_CMD\_PARSE\_UD} \(\backslash\)
                                 | \hyperlink{group__file__utils_gaf5e20210173cdb83bf70256a454f284b}{VRNA\_CMD\_PARSE\_SD} \(\backslash\)
                                 )
\end{DoxyCode}


Command parse/apply flag indicating default set of commands. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_structvrna__command__s}{vrna\+\_\+command\+\_\+s}, \hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\+\_\+file\+\_\+commands\+\_\+read()}, \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_gabec89c09874528c6cb73140a4c3d86d7}\label{group__file__utils_gabec89c09874528c6cb73140a4c3d86d7}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}}
\index{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}{VRNA\_OPTION\_MULTILINE}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+O\+P\+T\+I\+O\+N\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE~32U}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Tell a function that an input is assumed to span several lines. 

If used as input-\/option a function might also be returning this state telling that it has read data from multiple lines.

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure()}, \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga7d725ef525b29891abef3f1ed42599a4}\label{group__file__utils_ga7d725ef525b29891abef3f1ed42599a4}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}}
\index{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE@{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE}{VRNA\_CONSTRAINT\_MULTILINE}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+M\+U\+L\+T\+I\+L\+I\+NE~32U}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



parse multiline constraint 

\begin{DoxyRefDesc}{Deprecated}
\item[\hyperlink{deprecated__deprecated000065}{Deprecated}]see \hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure()} \end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}\label{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}{VRNA\_FILE\_FORMAT\_MSA\_CLUSTAL}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL~1U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating ClustalW formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}\label{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}{VRNA\_FILE\_FORMAT\_MSA\_STOCKHOLM}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM~2U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating Stockholm 1.\+0 formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_gacf6274a2c825f34e4131404665b00604}\label{group__file__utils_gacf6274a2c825f34e4131404665b00604}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}{VRNA\_FILE\_FORMAT\_MSA\_FASTA}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA~4U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating F\+A\+S\+TA (Pearson) formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_gab3186d774c01570d2a47d0308eec4927}\label{group__file__utils_gab3186d774c01570d2a47d0308eec4927}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}{VRNA\_FILE\_FORMAT\_MSA\_MAF}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF~8U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating M\+AF formatted files. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga494488a771aa0c602fb4cf445be34d47}\label{group__file__utils_ga494488a771aa0c602fb4cf445be34d47}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS}{VRNA\_FILE\_FORMAT\_MSA\_MIS}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS~16U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating most informative sequence (M\+IS) output. 

The default reference sequence output for an alignment is simply a consensus sequence. This flag allows to write the most informative equence (M\+IS) instead.

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga0c90faf1515ce0b16620036b1055ff1e}{vrna\+\_\+file\+\_\+msa\+\_\+write()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga4acc255373831856a8417b68de8a94c0}\label{group__file__utils_ga4acc255373831856a8417b68de8a94c0}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}{VRNA\_FILE\_FORMAT\_MSA\_DEFAULT}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}

{\bfseries Value\+:}
\begin{DoxyCode}
( \(\backslash\)
    VRNA\_FILE\_FORMAT\_MSA\_CLUSTAL \(\backslash\)
    | \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{VRNA\_FILE\_FORMAT\_MSA\_STOCKHOLM} \(\backslash\)
    | \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{VRNA\_FILE\_FORMAT\_MSA\_FASTA} \(\backslash\)
    | \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{VRNA\_FILE\_FORMAT\_MSA\_MAF} \(\backslash\)
    )
\end{DoxyCode}


Option flag indicating the set of default file formats. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}\label{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}{VRNA\_FILE\_FORMAT\_MSA\_NOCHECK}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK~4096U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag to disable validation of the alignment. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_gabdc948f547e550125de3e7c65878400c}\label{group__file__utils_gabdc948f547e550125de3e7c65878400c}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}{VRNA\_FILE\_FORMAT\_MSA\_UNKNOWN}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN~8192U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Return flag of \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} to indicate unknown or malformatted alignment. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga1577ea0f497d9c501549c863a4f2c089}\label{group__file__utils_ga1577ea0f497d9c501549c863a4f2c089}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND}{VRNA\_FILE\_FORMAT\_MSA\_APPEND}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND~16384U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag indicating to append data to a multiple sequence alignment file rather than overwriting it. 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga0c90faf1515ce0b16620036b1055ff1e}{vrna\+\_\+file\+\_\+msa\+\_\+write()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_ga73aa3701a22c5acfcb2e8927287a0554}\label{group__file__utils_ga73aa3701a22c5acfcb2e8927287a0554}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+Q\+U\+I\+ET@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+Q\+U\+I\+ET}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+Q\+U\+I\+ET@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+Q\+U\+I\+ET}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+Q\+U\+I\+ET}{VRNA\_FILE\_FORMAT\_MSA\_QUIET}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+Q\+U\+I\+ET~32768U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag to suppress unnecessary spam messages on {\ttfamily stderr} 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()} 
\end{DoxySeeAlso}
\mbox{\Hypertarget{group__file__utils_gad98ca08968d19f221f29d2a1faf26887}\label{group__file__utils_gad98ca08968d19f221f29d2a1faf26887}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+I\+L\+E\+NT@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+I\+L\+E\+NT}}
\index{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+I\+L\+E\+NT@{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+I\+L\+E\+NT}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+I\+L\+E\+NT}{VRNA\_FILE\_FORMAT\_MSA\_SILENT}}
{\footnotesize\ttfamily \#define V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+I\+L\+E\+NT~65536U}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Option flag to completely silence any warnings on {\ttfamily stderr} 

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()} 
\end{DoxySeeAlso}


\subsection{Function Documentation}
\mbox{\Hypertarget{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}\label{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+commands\+\_\+read@{vrna\+\_\+file\+\_\+commands\+\_\+read}}
\index{vrna\+\_\+file\+\_\+commands\+\_\+read@{vrna\+\_\+file\+\_\+commands\+\_\+read}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+commands\+\_\+read()}{vrna\_file\_commands\_read()}}
{\footnotesize\ttfamily \hyperlink{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}{vrna\+\_\+cmd\+\_\+t}$\ast$ vrna\+\_\+file\+\_\+commands\+\_\+read (\begin{DoxyParamCaption}\item[{const char $\ast$}]{filename,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Extract a list of commands from a command file. 

Read a list of commands specified in the input file and return them as list of abstract commands

\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}{vrna\+\_\+file\+\_\+commands\+\_\+apply()}, \hyperlink{group__file__utils_ga926fd7632e68f3da2e2069402b80b1ab}{vrna\+\_\+commands\+\_\+free()}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em filename} & The filename \\
\hline
{\em options} & Options to limit the type of commands read from the file \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
A list of abstract commands 
\end{DoxyReturn}
\mbox{\Hypertarget{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}\label{group__file__utils_gadbe8c9622f7bcc6dcbe3448b98df8656}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+commands\+\_\+apply@{vrna\+\_\+file\+\_\+commands\+\_\+apply}}
\index{vrna\+\_\+file\+\_\+commands\+\_\+apply@{vrna\+\_\+file\+\_\+commands\+\_\+apply}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+commands\+\_\+apply()}{vrna\_file\_commands\_apply()}}
{\footnotesize\ttfamily int vrna\+\_\+file\+\_\+commands\+\_\+apply (\begin{DoxyParamCaption}\item[{\hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$}]{vc,  }\item[{const char $\ast$}]{filename,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Apply a list of commands from a command file. 

This function is a shortcut to directly parse a commands file and apply all successfully parsed commands to a \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} data structure. It is the same as\+: 
\begin{DoxyCodeInclude}
  \textcolor{keywordtype}{int}         r;
  \hyperlink{group__file__utils_structvrna__command__s}{vrna\_cmd\_t}  *cmds;

  cmds  = \hyperlink{group__file__utils_ga5d2a64331cff5b1059e7d327545d8a63}{vrna\_file\_commands\_read}(filename, options);
  r     = \hyperlink{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}{vrna\_commands\_apply}(vc, cmds, options);

  \hyperlink{group__file__utils_ga926fd7632e68f3da2e2069402b80b1ab}{vrna\_commands\_free}(cmds);

  \textcolor{keywordflow}{return} r;
\end{DoxyCodeInclude}
 
\begin{DoxyParams}{Parameters}
{\em vc} & The \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} the command list will be applied to \\
\hline
{\em filename} & The filename \\
\hline
{\em options} & Options to limit the type of commands read from the file \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The number of commands successfully applied
\end{DoxyReturn}
\begin{DoxyRefDesc}{S\+W\+I\+G Wrapper Notes}
\item[\hyperlink{wrappers__wrappers000004}{S\+W\+I\+G Wrapper Notes}]This function is attached as method {\bfseries file\+\_\+commands\+\_\+apply()} to objects of type {\itshape fold\+\_\+compound} \end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}\label{group__file__utils_ga5e993fc4b9602af73aaaab4d3b3cd9a9}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+commands\+\_\+apply@{vrna\+\_\+commands\+\_\+apply}}
\index{vrna\+\_\+commands\+\_\+apply@{vrna\+\_\+commands\+\_\+apply}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+commands\+\_\+apply()}{vrna\_commands\_apply()}}
{\footnotesize\ttfamily int vrna\+\_\+commands\+\_\+apply (\begin{DoxyParamCaption}\item[{\hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} $\ast$}]{vc,  }\item[{\hyperlink{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}{vrna\+\_\+cmd\+\_\+t} $\ast$}]{commands,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Apply a list of commands to a \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t}. 


\begin{DoxyParams}{Parameters}
{\em vc} & The \hyperlink{group__fold__compound_ga1b0cef17fd40466cef5968eaeeff6166}{vrna\+\_\+fold\+\_\+compound\+\_\+t} the command list will be applied to \\
\hline
{\em commands} & The list of commands to apply \\
\hline
{\em options} & Options to limit the type of commands read from the file \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The number of commands successfully applied
\end{DoxyReturn}
\begin{DoxyRefDesc}{S\+W\+I\+G Wrapper Notes}
\item[\hyperlink{wrappers__wrappers000005}{S\+W\+I\+G Wrapper Notes}]This function is attached as method {\bfseries commands\+\_\+apply()} to objects of type {\itshape fold\+\_\+compound} \end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_ga926fd7632e68f3da2e2069402b80b1ab}\label{group__file__utils_ga926fd7632e68f3da2e2069402b80b1ab}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+commands\+\_\+free@{vrna\+\_\+commands\+\_\+free}}
\index{vrna\+\_\+commands\+\_\+free@{vrna\+\_\+commands\+\_\+free}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+commands\+\_\+free()}{vrna\_commands\_free()}}
{\footnotesize\ttfamily void vrna\+\_\+commands\+\_\+free (\begin{DoxyParamCaption}\item[{\hyperlink{group__file__utils_ga92cb3b5952352b103bcb32e5a99e0e5a}{vrna\+\_\+cmd\+\_\+t} $\ast$}]{commands }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{commands_8h}{Vienna\+R\+N\+A/commands.\+h}$>$}



Free memory occupied by a list of commands. 

Release memory occupied by a list of commands 
\begin{DoxyParams}{Parameters}
{\em commands} & A pointer to a list of commands \\
\hline
\end{DoxyParams}
\mbox{\Hypertarget{group__file__utils_gaaface7db12fadc3d271641c4515ab6e4}\label{group__file__utils_gaaface7db12fadc3d271641c4515ab6e4}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+helixlist@{vrna\+\_\+file\+\_\+helixlist}}
\index{vrna\+\_\+file\+\_\+helixlist@{vrna\+\_\+file\+\_\+helixlist}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+helixlist()}{vrna\_file\_helixlist()}}
{\footnotesize\ttfamily void vrna\+\_\+file\+\_\+helixlist (\begin{DoxyParamCaption}\item[{const char $\ast$}]{seq,  }\item[{const char $\ast$}]{db,  }\item[{float}]{energy,  }\item[{F\+I\+LE $\ast$}]{file }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure as helix list. 


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em energy} & Free energy of the structure in kcal/mol \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\mbox{\Hypertarget{group__file__utils_gab69682373ccca1e0e28cc967eec07745}\label{group__file__utils_gab69682373ccca1e0e28cc967eec07745}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+connect@{vrna\+\_\+file\+\_\+connect}}
\index{vrna\+\_\+file\+\_\+connect@{vrna\+\_\+file\+\_\+connect}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+connect()}{vrna\_file\_connect()}}
{\footnotesize\ttfamily void vrna\+\_\+file\+\_\+connect (\begin{DoxyParamCaption}\item[{const char $\ast$}]{seq,  }\item[{const char $\ast$}]{db,  }\item[{float}]{energy,  }\item[{const char $\ast$}]{identifier,  }\item[{F\+I\+LE $\ast$}]{file }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure as connect table. 

Connect table file format looks like this\+: \begin{DoxyVerb}300  ENERGY = 7.0  example
  1 G       0    2   22    1
  2 G       1    3   21    2
\end{DoxyVerb}
 where the headerline is followed by 6 columns with\+:
\begin{DoxyEnumerate}
\item Base number\+: index n
\item Base (A, C, G, T, U, X)
\item Index n-\/1 (0 if first nucleotide)
\item Index n+1 (0 if last nucleotide)
\item Number of the base to which n is paired. No pairing is indicated by 0 (zero).
\item Natural numbering.
\end{DoxyEnumerate}


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em energy} & The free energy of the structure \\
\hline
{\em identifier} & An optional identifier for the sequence \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\mbox{\Hypertarget{group__file__utils_ga9b462e6f202594af5d3fa56e280d633f}\label{group__file__utils_ga9b462e6f202594af5d3fa56e280d633f}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+bpseq@{vrna\+\_\+file\+\_\+bpseq}}
\index{vrna\+\_\+file\+\_\+bpseq@{vrna\+\_\+file\+\_\+bpseq}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+bpseq()}{vrna\_file\_bpseq()}}
{\footnotesize\ttfamily void vrna\+\_\+file\+\_\+bpseq (\begin{DoxyParamCaption}\item[{const char $\ast$}]{seq,  }\item[{const char $\ast$}]{db,  }\item[{F\+I\+LE $\ast$}]{file }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure in bpseq format. 


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\mbox{\Hypertarget{group__file__utils_ga31f4a6c2ea1495a6e4f9eb45a9f6193d}\label{group__file__utils_ga31f4a6c2ea1495a6e4f9eb45a9f6193d}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+json@{vrna\+\_\+file\+\_\+json}}
\index{vrna\+\_\+file\+\_\+json@{vrna\+\_\+file\+\_\+json}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+json()}{vrna\_file\_json()}}
{\footnotesize\ttfamily void vrna\+\_\+file\+\_\+json (\begin{DoxyParamCaption}\item[{const char $\ast$}]{seq,  }\item[{const char $\ast$}]{db,  }\item[{double}]{energy,  }\item[{const char $\ast$}]{identifier,  }\item[{F\+I\+LE $\ast$}]{file }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Print a secondary structure in jsonformat. 


\begin{DoxyParams}{Parameters}
{\em seq} & The R\+NA sequence \\
\hline
{\em db} & The structure in dot-\/bracket format \\
\hline
{\em energy} & The free energy \\
\hline
{\em identifier} & An identifier for the sequence \\
\hline
{\em file} & The file handle used to print to (print defaults to \textquotesingle{}stdout\textquotesingle{} if(file == N\+U\+LL) ) \\
\hline
\end{DoxyParams}
\mbox{\Hypertarget{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}\label{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record}}
\index{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()}{vrna\_file\_fasta\_read\_record()}}
{\footnotesize\ttfamily unsigned int vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record (\begin{DoxyParamCaption}\item[{char $\ast$$\ast$}]{header,  }\item[{char $\ast$$\ast$}]{sequence,  }\item[{char $\ast$$\ast$$\ast$}]{rest,  }\item[{F\+I\+LE $\ast$}]{file,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Get a (fasta) data set from a file or stdin. 

This function may be used to obtain complete datasets from a filehandle or stdin. A dataset is always defined to contain at least a sequence. If data starts with a fasta header, i.\+e. a line like \begin{DoxyVerb}>some header info \end{DoxyVerb}
 then \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} will assume that the sequence that follows the header may span over several lines. To disable this behavior and to assign a single line to the argument \textquotesingle{}sequence\textquotesingle{} one can pass \hyperlink{group__utils_ga0de536599b881c787b0943a2671da476}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+\_\+\+S\+P\+AN} in the \textquotesingle{}options\textquotesingle{} argument. If no fasta header is read in the beginning of a data block, a sequence must not span over multiple lines!~\newline
 Unless the options \hyperlink{group__utils_ga0f6311f11bed1842e3a527ab27b294c6}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+S\+K\+I\+P\+\_\+\+C\+O\+M\+M\+E\+N\+TS} or \hyperlink{group__utils_gab4db885222b3b69608310d7c7e63e286}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+S\+K\+I\+P\+\_\+\+B\+L\+A\+N\+K\+\_\+\+L\+I\+N\+ES} are passed, a sequence may be interrupted by lines starting with a comment character or empty lines.~\newline
 A sequence is regarded as completely read if it was either assumed to not span over multiple lines, a secondary structure or structure constraint follows the sequence on the next line, or a new header marks the beginning of a new sequence...~\newline
 All lines following the sequence (this includes comments) that do not initiate a new dataset according to the above definition are available through the line-\/array \textquotesingle{}rest\textquotesingle{}. Here one can usually find the structure constraint or other information belonging to the current dataset. Filling of \textquotesingle{}rest\textquotesingle{} may be prevented by passing \hyperlink{group__utils_ga7a2e8c50a0c7ce82e60da1016e1367fd}{V\+R\+N\+A\+\_\+\+I\+N\+P\+U\+T\+\_\+\+N\+O\+\_\+\+R\+E\+ST} to the options argument.~\newline
 \begin{DoxyNote}{Note}
This function will exit any program with an error message if no sequence could be read! 

This function is N\+OT threadsafe! It uses a global variable to store information about the next data block.
\end{DoxyNote}
The main purpose of this function is to be able to easily parse blocks of data in the header of a loop where all calculations for the appropriate data is done inside the loop. The loop may be then left on certain return values, e.\+g.\+: 
\begin{DoxyCode}
\textcolor{keywordtype}{char} *id, *seq, **rest;
\textcolor{keywordtype}{int}  i;
\textcolor{keywordtype}{id} = seq = NULL;
rest = NULL;
\textcolor{keywordflow}{while}(!(\hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\_file\_fasta\_read\_record}(&\textcolor{keywordtype}{id}, &seq, &rest, NULL, 0) & (
      \hyperlink{group__utils_gad403c9ea58f1836689404c2931419c8c}{VRNA\_INPUT\_ERROR} | \hyperlink{group__utils_ga72f3c6ca5c83d2b9baed2922d19c403d}{VRNA\_INPUT\_QUIT})))\{
  \textcolor{keywordflow}{if}(\textcolor{keywordtype}{id}) printf(\textcolor{stringliteral}{"%s\(\backslash\)n"}, \textcolor{keywordtype}{id});
  printf(\textcolor{stringliteral}{"%s\(\backslash\)n"}, seq);
  \textcolor{keywordflow}{if}(rest)
    \textcolor{keywordflow}{for}(i=0;rest[i];i++)\{
      printf(\textcolor{stringliteral}{"%s\(\backslash\)n"}, rest[i]);
      free(rest[i]);
    \}
  free(rest);
  free(seq);
  free(\textcolor{keywordtype}{id});
\}
\end{DoxyCode}
 In the example above, the while loop will be terminated when \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} returns either an error, E\+OF, or a user initiated quit request.~\newline
 As long as data is read from stdin (we are passing N\+U\+LL as the file pointer), the id is printed if it is available for the current block of data. The sequence will be printed in any case and if some more lines belong to the current block of data each line will be printed as well.

\begin{DoxyNote}{Note}
Do not forget to free the memory occupied by header, sequence and rest!
\end{DoxyNote}

\begin{DoxyParams}{Parameters}
{\em header} & A pointer which will be set such that it points to the header of the record \\
\hline
{\em sequence} & A pointer which will be set such that it points to the sequence of the record \\
\hline
{\em rest} & A pointer which will be set such that it points to an array of lines which also belong to the record \\
\hline
{\em file} & A file handle to read from (if N\+U\+LL, this function reads from stdin) \\
\hline
{\em options} & Some options which may be passed to alter the behavior of the function, use 0 for no options \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
A flag with information about what the function actually did read 
\end{DoxyReturn}
\mbox{\Hypertarget{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}\label{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure}}
\index{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure()}{vrna\_extract\_record\_rest\_structure()}}
{\footnotesize\ttfamily char$\ast$ vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure (\begin{DoxyParamCaption}\item[{const char $\ast$$\ast$}]{lines,  }\item[{unsigned int}]{length,  }\item[{unsigned int}]{option }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Extract a dot-\/bracket structure string from (multiline)character array. 

This function extracts a dot-\/bracket structure string from the \textquotesingle{}rest\textquotesingle{} array as returned by \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} and returns it. All occurences of comments within the \textquotesingle{}lines\textquotesingle{} array will be skipped as long as they do not break the structure string. If no structure could be read, this function returns N\+U\+LL.

\begin{DoxyPrecond}{Precondition}
The argument \textquotesingle{}lines\textquotesingle{} has to be a 2-\/dimensional character array as obtained by \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} 
\end{DoxyPrecond}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em lines} & The (multiline) character array to be parsed \\
\hline
{\em length} & The assumed length of the dot-\/bracket string (passing a value $<$ 1 results in no length limit) \\
\hline
{\em option} & Some options which may be passed to alter the behavior of the function, use 0 for no options \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The dot-\/bracket string read from lines or N\+U\+LL 
\end{DoxyReturn}
\mbox{\Hypertarget{group__file__utils_ga646ebf45450a69a7f2533f9ecd283a32}\label{group__file__utils_ga646ebf45450a69a7f2533f9ecd283a32}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read@{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read}}
\index{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read@{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read()}{vrna\_file\_SHAPE\_read()}}
{\footnotesize\ttfamily int vrna\+\_\+file\+\_\+\+S\+H\+A\+P\+E\+\_\+read (\begin{DoxyParamCaption}\item[{const char $\ast$}]{file\+\_\+name,  }\item[{int}]{length,  }\item[{double}]{default\+\_\+value,  }\item[{char $\ast$}]{sequence,  }\item[{double $\ast$}]{values }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Read data from a given S\+H\+A\+PE reactivity input file. 

This function parses the informations from a given file and stores the result in the preallocated string sequence and the double array values.


\begin{DoxyParams}{Parameters}
{\em file\+\_\+name} & Path to the constraints file \\
\hline
{\em length} & Length of the sequence (file entries exceeding this limit will cause an error) \\
\hline
{\em default\+\_\+value} & Value for missing indices \\
\hline
{\em sequence} & Pointer to an array used for storing the sequence obtained from the S\+H\+A\+PE reactivity file \\
\hline
{\em values} & Pointer to an array used for storing the values obtained from the S\+H\+A\+PE reactivity file \\
\hline
\end{DoxyParams}
\mbox{\Hypertarget{group__file__utils_ga55a9ae6dfeecc1b3f0c2acf6fa796c15}\label{group__file__utils_ga55a9ae6dfeecc1b3f0c2acf6fa796c15}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint}}
\index{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint@{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint()}{vrna\_extract\_record\_rest\_constraint()}}
{\footnotesize\ttfamily void vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+constraint (\begin{DoxyParamCaption}\item[{char $\ast$$\ast$}]{cstruc,  }\item[{const char $\ast$$\ast$}]{lines,  }\item[{unsigned int}]{option }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Extract a hard constraint encoded as pseudo dot-\/bracket string. 

\begin{DoxyRefDesc}{Deprecated}
\item[\hyperlink{deprecated__deprecated000066}{Deprecated}]Use \hyperlink{group__file__utils_gad37cbb63a05eed63ba25c91628409be0}{vrna\+\_\+extract\+\_\+record\+\_\+rest\+\_\+structure()} instead! \end{DoxyRefDesc}
\begin{DoxyPrecond}{Precondition}
The argument \textquotesingle{}lines\textquotesingle{} has to be a 2-\/dimensional character array as obtained by \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} 
\end{DoxyPrecond}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()}, \hyperlink{group__hard__constraints_ga13053547a2de5532b64b64d35e097ae1}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+P\+I\+PE}, \hyperlink{group__hard__constraints_ga369bea82eae75fbe626f409fa425747e}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+D\+OT}, \hyperlink{group__hard__constraints_ga7283bbe0f8954f7b030ecc3f2d1932b2}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+X} \hyperlink{constraints__hard_8h_ad54c1315a47d55653dcaa5de6e544b77}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+A\+N\+G\+\_\+\+B\+R\+A\+CK}, \hyperlink{group__hard__constraints_gac17b034852c914bc5879954c65d7e74b}{V\+R\+N\+A\+\_\+\+C\+O\+N\+S\+T\+R\+A\+I\+N\+T\+\_\+\+D\+B\+\_\+\+R\+N\+D\+\_\+\+B\+R\+A\+CK}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em cstruc} & A pointer to a character array that is used as pseudo dot-\/bracket output \\
\hline
{\em lines} & A 2-\/dimensional character array with the extension lines from the F\+A\+S\+TA input \\
\hline
{\em option} & The option flags that define the behavior and recognition pattern of this function \\
\hline
\end{DoxyParams}
\mbox{\Hypertarget{group__file__utils_gafd194a69af9d92b5b0412a7627ac1595}\label{group__file__utils_gafd194a69af9d92b5b0412a7627ac1595}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!read\+\_\+record@{read\+\_\+record}}
\index{read\+\_\+record@{read\+\_\+record}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{read\+\_\+record()}{read\_record()}}
{\footnotesize\ttfamily unsigned int read\+\_\+record (\begin{DoxyParamCaption}\item[{char $\ast$$\ast$}]{header,  }\item[{char $\ast$$\ast$}]{sequence,  }\item[{char $\ast$$\ast$$\ast$}]{rest,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats_8h}{Vienna\+R\+N\+A/file\+\_\+formats.\+h}$>$}



Get a data record from stdin. 

\begin{DoxyRefDesc}{Deprecated}
\item[\hyperlink{deprecated__deprecated000068}{Deprecated}]This function is deprecated! Use \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()} as a replacment.\end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}\label{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+msa\+\_\+read@{vrna\+\_\+file\+\_\+msa\+\_\+read}}
\index{vrna\+\_\+file\+\_\+msa\+\_\+read@{vrna\+\_\+file\+\_\+msa\+\_\+read}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+msa\+\_\+read()}{vrna\_file\_msa\_read()}}
{\footnotesize\ttfamily vrna\+\_\+file\+\_\+msa\+\_\+read (\begin{DoxyParamCaption}\item[{const char $\ast$}]{filename,  }\item[{char $\ast$$\ast$$\ast$}]{names,  }\item[{char $\ast$$\ast$$\ast$}]{aln,  }\item[{char $\ast$$\ast$}]{id,  }\item[{char $\ast$$\ast$}]{structure,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Read a multiple sequence alignment from file. 

This function reads the (first) multiple sequence alignment from an input file. The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The {\ttfamily options} parameter allows to specify the set of alignment file formats that should be used to retrieve the data. If 0 is passed as option, the list of alignment file formats defaults to \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}.

Currently, the list of parsable multiple sequence alignment file formats consists of\+:
\begin{DoxyItemize}
\item \hyperlink{file_formats_msa-formats-clustal}{ClustalW format}
\item \hyperlink{file_formats_msa-formats-stockholm}{Stockholm 1.\+0 format}
\item \hyperlink{file_formats_msa-formats-fasta}{F\+A\+S\+TA (Pearson) format}
\item \hyperlink{file_formats_msa-formats-maf}{M\+AF format}
\end{DoxyItemize}\begin{DoxyNote}{Note}
After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK} in the {\ttfamily options} parameter.

It is the users responsibility to free any memory occupied by the output arguments {\ttfamily names}, {\ttfamily aln}, {\ttfamily id}, and {\ttfamily structure} after calling this function. The function automatically sets the latter two arguments to {\ttfamily N\+U\+LL} in case no corresponding data could be retrieved from the input alignment.
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}, \hyperlink{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}, \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}, \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}, \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}, \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}, \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em filename} & The name of input file that contains the alignment \\
\hline
{\em names} & An address to the pointer where sequence identifiers should be written to \\
\hline
{\em aln} & An address to the pointer where aligned sequences should be written to \\
\hline
{\em id} & An address to the pointer where the alignment ID should be written to (Maybe N\+U\+LL) \\
\hline
{\em structure} & An address to the pointer where consensus structure information should be written to (Maybe N\+U\+LL) \\
\hline
{\em options} & Options to manipulate the behavior of this function \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The number of sequences in the alignment, or -\/1 if no alignment record could be found
\end{DoxyReturn}
\begin{DoxyRefDesc}{S\+W\+I\+G Wrapper Notes}
\item[\hyperlink{wrappers__wrappers000060}{S\+W\+I\+G Wrapper Notes}]In the target scripting language, only the first and last argument, {\ttfamily filename} and {\ttfamily options}, are passed to the corresponding function. The other arguments, which serve as output in the C-\/library, are available as additional return values. Hence, a function call in python may look like this\+:


\begin{DoxyCode}
num\_seq, names, aln, id, structure = RNA.file\_msa\_read(\textcolor{stringliteral}{"msa.stk"}, RNA.FILE\_FORMAT\_MSA\_STOCKHOLM)
\end{DoxyCode}


After successfully reading the first record, the variable {\ttfamily num\+\_\+seq} contains the number of sequences in the alignment (the actual return value of the C-\/function), while the variables {\ttfamily names}, {\ttfamily aln}, {\ttfamily id}, and {\ttfamily structure} are lists of the sequence names and aligned sequences, as well as strings holding the alignment ID and the structure as stated in the {\ttfamily S\+S\+\_\+cons} line, respectively. Note, the last two return values may be empty strings in case the alignment does not provide the required data.

This function exists as an overloaded version where the {\ttfamily options} parameter may be omitted! In that case, the {\ttfamily options} parameter defaults to \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}. \end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}\label{group__file__utils_ga59204cd1daa4927f5127cc65a2886efd}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record}}
\index{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record@{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record()}{vrna\_file\_msa\_read\_record()}}
{\footnotesize\ttfamily vrna\+\_\+file\+\_\+msa\+\_\+read\+\_\+record (\begin{DoxyParamCaption}\item[{F\+I\+LE $\ast$}]{fp,  }\item[{char $\ast$$\ast$$\ast$}]{names,  }\item[{char $\ast$$\ast$$\ast$}]{aln,  }\item[{char $\ast$$\ast$}]{id,  }\item[{char $\ast$$\ast$}]{structure,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Read a multiple sequence alignment from file handle. 

Similar to \hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, this function reads a multiple sequence alignment from an input file handle. Since using a file handle, this function is not limited to the first alignment record, but allows for looping over all alignments within the input.

The read alignment is split into the sequence id/name part and the actual sequence information and stored in memory as arrays of ids/names and sequences. If the alignment file format allows for additional information, such as an ID of the entire alignment or consensus structure information, this data is retrieved as well and made available. The {\ttfamily options} parameter allows to specify the alignment file format used to retrieve the data. A single format must be specified here, see \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()} for helping to determine the correct M\+SA file format.

Currently, the list of parsable multiple sequence alignment file formats consists of\+:
\begin{DoxyItemize}
\item \hyperlink{file_formats_msa-formats-clustal}{ClustalW format}
\item \hyperlink{file_formats_msa-formats-stockholm}{Stockholm 1.\+0 format}
\item \hyperlink{file_formats_msa-formats-fasta}{F\+A\+S\+TA (Pearson) format}
\item \hyperlink{file_formats_msa-formats-maf}{M\+AF format}
\end{DoxyItemize}\begin{DoxyNote}{Note}
After successfully reading an alignment, this function performs a validation of the data that includes uniqueness of the sequence identifiers, and equal sequence lengths. This check can be deactivated by passing \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK} in the {\ttfamily options} parameter.

It is the users responsibility to free any memory occupied by the output arguments {\ttfamily names}, {\ttfamily aln}, {\ttfamily id}, and {\ttfamily structure} after calling this function. The function automatically sets the latter two arguments to {\ttfamily N\+U\+LL} in case no corresponding data could be retrieved from the input alignment.
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, \hyperlink{group__file__utils_ga627ac281b5f11c63861726e6472626c9}{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()}, \hyperlink{group__file__utils_ga79a23de2c7249f2cccd762e475c81859}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+C\+L\+U\+S\+T\+AL}, \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}, \hyperlink{group__file__utils_gacf6274a2c825f34e4131404665b00604}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+F\+A\+S\+TA}, \hyperlink{group__file__utils_gab3186d774c01570d2a47d0308eec4927}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+AF}, \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}, \hyperlink{group__file__utils_ga229fb3778ecabea4782902b69fa48fd1}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+N\+O\+C\+H\+E\+CK}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em fp} & The file pointer the data will be retrieved from \\
\hline
{\em names} & An address to the pointer where sequence identifiers should be written to \\
\hline
{\em aln} & An address to the pointer where aligned sequences should be written to \\
\hline
{\em id} & An address to the pointer where the alignment ID should be written to (Maybe N\+U\+LL) \\
\hline
{\em structure} & An address to the pointer where consensus structure information should be written to (Maybe N\+U\+LL) \\
\hline
{\em options} & Options to manipulate the behavior of this function \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The number of sequences in the alignment, or -\/1 if no alignment record could be found
\end{DoxyReturn}
\begin{DoxyRefDesc}{S\+W\+I\+G Wrapper Notes}
\item[\hyperlink{wrappers__wrappers000061}{S\+W\+I\+G Wrapper Notes}]In the target scripting language, only the first and last argument, {\ttfamily fp} and {\ttfamily options}, are passed to the corresponding function. The other arguments, which serve as output in the C-\/library, are available as additional return values. Hence, a function call in python may look like this\+:


\begin{DoxyCode}
f = open(\textcolor{stringliteral}{'msa.stk'}, \textcolor{charliteral}{'r'})
num\_seq, names, aln, \textcolor{keywordtype}{id}, structure = RNA.file\_msa\_read\_record(f, RNA.FILE\_FORMAT\_MSA\_STOCKHOLM)
f.close() 
\end{DoxyCode}


After successfully reading the first record, the variable {\ttfamily num\+\_\+seq} contains the number of sequences in the alignment (the actual return value of the C-\/function), while the variables {\ttfamily names}, {\ttfamily aln}, {\ttfamily id}, and {\ttfamily structure} are lists of the sequence names and aligned sequences, as well as strings holding the alignment ID and the structure as stated in the {\ttfamily S\+S\+\_\+cons} line, respectively. Note, the last two return values may be empty strings in case the alignment does not provide the required data.

This function exists as an overloaded version where the {\ttfamily options} parameter may be omitted! In that case, the {\ttfamily options} parameter defaults to \hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}. \end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_ga627ac281b5f11c63861726e6472626c9}\label{group__file__utils_ga627ac281b5f11c63861726e6472626c9}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format@{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format}}
\index{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format@{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format()}{vrna\_file\_msa\_detect\_format()}}
{\footnotesize\ttfamily vrna\+\_\+file\+\_\+msa\+\_\+detect\+\_\+format (\begin{DoxyParamCaption}\item[{const char $\ast$}]{fn,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Detect the format of a multiple sequence alignment file. 

This function attempts to determine the format of a file that supposedly contains a multiple sequence alignment (M\+SA). This is useful in cases where a M\+SA file contains more than a single record and therefore \hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()} can not be applied, since it only retrieves the first. Here, one can try to guess the correct file format using this function and then loop over the file, record by record using one of the low-\/level record retrieval functions for the corresponding M\+SA file format.

\begin{DoxyNote}{Note}
This function parses the entire first record within the specified file. As a result, it returns \hyperlink{group__file__utils_gabdc948f547e550125de3e7c65878400c}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN} not only if it can\textquotesingle{}t detect the file\textquotesingle{}s format, but also in cases where the file doesn\textquotesingle{}t contain sequences!
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga08a01c40ac5f5e0e04e9ae2258c99aa6}{vrna\+\_\+file\+\_\+msa\+\_\+read()}, vrna\+\_\+file\+\_\+stockholm\+\_\+read\+\_\+record(), vrna\+\_\+file\+\_\+clustal\+\_\+read\+\_\+record(), \hyperlink{group__file__utils_ga8cfb7e271efc9e1f34640acb85475639}{vrna\+\_\+file\+\_\+fasta\+\_\+read\+\_\+record()}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em filename} & The name of input file that contains the alignment \\
\hline
{\em options} & Options to manipulate the behavior of this function \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The M\+SA file format, or \hyperlink{group__file__utils_gabdc948f547e550125de3e7c65878400c}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+U\+N\+K\+N\+O\+WN}
\end{DoxyReturn}
\begin{DoxyRefDesc}{S\+W\+I\+G Wrapper Notes}
\item[\hyperlink{wrappers__wrappers000059}{S\+W\+I\+G Wrapper Notes}]This function exists as an overloaded version where the {\ttfamily options} parameter may be omitted! In that case, the {\ttfamily options} parameter defaults to \hyperlink{group__file__utils_ga4acc255373831856a8417b68de8a94c0}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+D\+E\+F\+A\+U\+LT}. \end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_ga0c90faf1515ce0b16620036b1055ff1e}\label{group__file__utils_ga0c90faf1515ce0b16620036b1055ff1e}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+file\+\_\+msa\+\_\+write@{vrna\+\_\+file\+\_\+msa\+\_\+write}}
\index{vrna\+\_\+file\+\_\+msa\+\_\+write@{vrna\+\_\+file\+\_\+msa\+\_\+write}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+file\+\_\+msa\+\_\+write()}{vrna\_file\_msa\_write()}}
{\footnotesize\ttfamily vrna\+\_\+file\+\_\+msa\+\_\+write (\begin{DoxyParamCaption}\item[{const char $\ast$}]{filename,  }\item[{const char $\ast$$\ast$}]{names,  }\item[{const char $\ast$$\ast$}]{aln,  }\item[{const char $\ast$}]{id,  }\item[{const char $\ast$}]{structure,  }\item[{const char $\ast$}]{source,  }\item[{unsigned int}]{options }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__formats__msa_8h}{Vienna\+R\+N\+A/file\+\_\+formats\+\_\+msa.\+h}$>$}



Write multiple sequence alignment file. 

\begin{DoxyNote}{Note}
Currently, we only support \hyperlink{file_formats_msa-formats-stockholm}{Stockholm 1.\+0 format} output
\end{DoxyNote}
\begin{DoxySeeAlso}{See also}
\hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM}, \hyperlink{group__file__utils_ga1577ea0f497d9c501549c863a4f2c089}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND}, \hyperlink{group__file__utils_ga494488a771aa0c602fb4cf445be34d47}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+M\+IS}
\end{DoxySeeAlso}

\begin{DoxyParams}{Parameters}
{\em filename} & The output filename \\
\hline
{\em names} & The array of sequence names / identifies \\
\hline
{\em aln} & The array of aligned sequences \\
\hline
{\em id} & An optional ID for the alignment \\
\hline
{\em structure} & An optional consensus structure \\
\hline
{\em source} & A string describing the source of the alignment \\
\hline
{\em options} & Options to manipulate the behavior of this function \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
Non-\/null upon successfully writing the alignment to file
\end{DoxyReturn}
\begin{DoxyRefDesc}{S\+W\+I\+G Wrapper Notes}
\item[\hyperlink{wrappers__wrappers000062}{S\+W\+I\+G Wrapper Notes}]In the target scripting language, this function exists as a set of overloaded versions, where the last four parameters may be omitted. If the {\ttfamily options} parameter is missing the options default to (\hyperlink{group__file__utils_ga62be992445cd8ab2ad7a8fded944338b}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+S\+T\+O\+C\+K\+H\+O\+LM} $\vert$ \hyperlink{group__file__utils_ga1577ea0f497d9c501549c863a4f2c089}{V\+R\+N\+A\+\_\+\+F\+I\+L\+E\+\_\+\+F\+O\+R\+M\+A\+T\+\_\+\+M\+S\+A\+\_\+\+A\+P\+P\+E\+ND}). \end{DoxyRefDesc}
\mbox{\Hypertarget{group__file__utils_gad16c270b70a77c753088b29de8a802dc}\label{group__file__utils_gad16c270b70a77c753088b29de8a802dc}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+read\+\_\+line@{vrna\+\_\+read\+\_\+line}}
\index{vrna\+\_\+read\+\_\+line@{vrna\+\_\+read\+\_\+line}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+read\+\_\+line()}{vrna\_read\_line()}}
{\footnotesize\ttfamily char$\ast$ vrna\+\_\+read\+\_\+line (\begin{DoxyParamCaption}\item[{F\+I\+LE $\ast$}]{fp }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__utils_8h}{Vienna\+R\+N\+A/file\+\_\+utils.\+h}$>$}



Read a line of arbitrary length from a stream. 

Returns a pointer to the resulting string. The necessary memory is allocated and should be released using {\itshape free()} when the string is no longer needed.


\begin{DoxyParams}{Parameters}
{\em fp} & A file pointer to the stream where the function should read from \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
A pointer to the resulting string 
\end{DoxyReturn}
\mbox{\Hypertarget{group__file__utils_ga72eae0a472bad905b6facccf5b453a76}\label{group__file__utils_ga72eae0a472bad905b6facccf5b453a76}} 
\index{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}!vrna\+\_\+filename\+\_\+sanitize@{vrna\+\_\+filename\+\_\+sanitize}}
\index{vrna\+\_\+filename\+\_\+sanitize@{vrna\+\_\+filename\+\_\+sanitize}!Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments@{Functions to Read/\+Write several File Formats for R\+N\+A Sequences, Structures, and Alignments}}
\subsubsection{\texorpdfstring{vrna\+\_\+filename\+\_\+sanitize()}{vrna\_filename\_sanitize()}}
{\footnotesize\ttfamily char$\ast$ vrna\+\_\+filename\+\_\+sanitize (\begin{DoxyParamCaption}\item[{const char $\ast$}]{name,  }\item[{const char $\ast$}]{replacement }\end{DoxyParamCaption})}



{\ttfamily \#include $<$\hyperlink{file__utils_8h}{Vienna\+R\+N\+A/file\+\_\+utils.\+h}$>$}



Sanitize a file name. 

Returns a new file name where all invalid characters are substituted by a replacement character. If no replacement character is supplied, invalid characters are simply removed from the filename. File names may also never exceed a length of 255 characters. Longer file names will undergo a \textquotesingle{}smart\textquotesingle{} truncation process, where the filenames` suffix, i.\+e. everything after the last dot \textquotesingle{}.\textquotesingle{}, is attempted to be kept intact. Hence, only the filename part before the suffix is reduced in such a way that the total filename complies to the length restriction of 255 characters. If no suffix is present or the suffix itself already exceeds the maximum length, the filename is simply truncated from the back of the string.

For now we consider the following characters invalid\+:
\begin{DoxyItemize}
\item backslash \textquotesingle{}\textbackslash{}\textquotesingle{}
\item slash \textquotesingle{}/\textquotesingle{}
\item question mark \textquotesingle{}?\textquotesingle{}
\item percent sign \textquotesingle{}\textquotesingle{}
\item asterisk \textquotesingle{}$\ast$\textquotesingle{}
\item colon \textquotesingle{}\+:\textquotesingle{}
\item pipe symbol \textquotesingle{}$\vert$\textquotesingle{}
\item double quote \textquotesingle{}\char`\"{}\textquotesingle{}
\item triangular brackets \textquotesingle{}$<$\textquotesingle{} and \textquotesingle{}$>$\textquotesingle{}
\end{DoxyItemize}

Furthermore, the (resulting) file name must not be a reserved file name, such as\+:
\begin{DoxyItemize}
\item \textquotesingle{}.\textquotesingle{}
\item \textquotesingle{}..\textquotesingle{}
\end{DoxyItemize}

\begin{DoxyNote}{Note}
This function allocates a new block of memory for the sanitized string. It also may return (a) N\+U\+LL if the input is pointing to N\+U\+LL, or (b) an empty string if the input only consists of invalid characters which are simply removed!
\end{DoxyNote}

\begin{DoxyParams}{Parameters}
{\em name} & The input file name \\
\hline
{\em replacment} & The replacement character, or N\+U\+LL \\
\hline
\end{DoxyParams}
\begin{DoxyReturn}{Returns}
The sanitized file name, or N\+U\+LL 
\end{DoxyReturn}
